{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rGreat local vs greatpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook allows you to compare results between local rGreat and enrichment function from greatpy. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "r-base == 3.6.1\n",
    "\n",
    "python version == 3.8 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects import r as r\n",
    "pandas2ri.activate()\n",
    "import pandas as pd\n",
    "\n",
    "import os \n",
    "import matplotlib.pyplot as plt\n",
    "import greatpy as great \n",
    "from scipy.stats import pearsonr\n",
    "import re\n",
    "import time "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.5.2'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpy2.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "importr('rGREAT')\n",
    "ranges = importr('GenomicRanges')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rgreat_online_vs_local_vs_greatpy(): \n",
    "    stat_df = {\n",
    "        \"name\" : [],\n",
    "        \"pearson_binom\" : [], \n",
    "        \"pearson_hypergeom\" : []\n",
    "    }\n",
    "    pp = {\n",
    "        \"name\" : [],\n",
    "        \"before_pp_greatpy_size\" : [], \n",
    "        \"before_pp_local_size\" : [],\n",
    "        \"final_size\" : [],\n",
    "        \"%_of_diffrent_GO_term\" : [],\n",
    "    }\n",
    "    # test_file = [\"01_random.bed\",\"06_height_snps_hg38.bed\",\"07_height_snps_hg19.bed\",\"10_MAX.bed\"]\n",
    "    test_file = [\"06_height_snps_hg38.bed\"]\n",
    "\n",
    "    for name in test_file : \n",
    "        # find the assembly \n",
    "        if re.match(\".*hg19.*\",name) != None : \n",
    "            assembly = \"hg19\"\n",
    "        else : \n",
    "            assembly = \"hg38\"\n",
    "\n",
    "        # online test \n",
    "        res_online = rpy2.robjects.r['submitGreatJob'](f\"../../data/tests/test_data/input/{name}\",species=f\"{assembly}\",help=False)\n",
    "        res_online = rpy2.robjects.r['getEnrichmentTables'](res_online)\n",
    "\n",
    "        time.sleep(30)\n",
    "\n",
    "        # local test\n",
    "            # proprocessing : make a Grange frame \n",
    "        df = r[\"read.csv\"](f\"../../data/tests/test_data/input/{name}\",sep='\\t')\n",
    "        seqname = rpy2.robjects.StrVector([\"seqnames\", \"seqname\",\"chromosome\", \"X.Chr\",\"chr\", \"chromosome_name\",\"seqid\"])\n",
    "        end = rpy2.robjects.StrVector(['end', 'stop'])\n",
    "        df = ranges.makeGRangesFromDataFrame(df,seqnames_field=seqname)\n",
    "\n",
    "            # great calculation\n",
    "        local = rpy2.robjects.r['great'](df, \"msigdb:C5\", f\"txdb:{assembly}\",verbose=False)\n",
    "        local = rpy2.robjects.r['getEnrichmentTables'](local)\n",
    "\n",
    "        time.sleep(30)\n",
    "\n",
    "        # greatpy calculation\n",
    "        greatpy = great.tl.GREAT.enrichment(\n",
    "            test_file=f\"../../data/tests/test_data/input/{name}\",\n",
    "            regdom_file=f\"../../data/human/{assembly}/regulatory_domain.bed\",\n",
    "            chr_size_file=f\"../../data/human/{assembly}/chr_size.bed\",\n",
    "            annotation_file=f\"../../data/human/ontologies.csv\",\n",
    "            binom=True,\n",
    "            hypergeom=True,\n",
    "            )\n",
    "        time.sleep(30)\n",
    "\n",
    "        # create each dataframe\n",
    "            # online\n",
    "        name = [cdc.lower() for cdc in list(res_online.rx2(\"GO Molecular Function\").rx2(\"name\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"name\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"name\"))]\n",
    "        online = pd.DataFrame({\n",
    "            \"id\" : list(res_online.rx2(\"GO Molecular Function\").rx2(\"ID\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"ID\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"ID\")),\n",
    "            \"name\" : name,\n",
    "            \"binom_p_val\" : list(res_online.rx2(\"GO Molecular Function\").rx2(\"Binom_Raw_PValue\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"Binom_Raw_PValue\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"Binom_Raw_PValue\")),\n",
    "            \"hyper_p_val\" : list(res_online.rx2(\"GO Molecular Function\").rx2(\"Hyper_Raw_PValue\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"Hyper_Raw_PValue\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"Hyper_Raw_PValue\"))\n",
    "            })\n",
    "\n",
    "            # local\n",
    "        name_go = list(local.rx2(\"id\")) \n",
    "        name_go = [\" \".join(cdc.lower().split(\"_\")[1:]) for cdc in list(local.rx2(\"id\"))]\n",
    "        local = pd.DataFrame({\n",
    "            \"name\": name_go,\n",
    "            \"binom_p_val\" : list(local.rx2(\"p_value\")),\n",
    "            \"hyper_p_val\" : list(local.rx2(\"p_value_hyper\"))\n",
    "            })\n",
    "\n",
    "            # greatpy\n",
    "        greatpy[\"go_term\"] = greatpy[\"go_term\"].str.lower()\n",
    "\n",
    "        \n",
    "        time.sleep(90) # just to have not a burning computer ^^ \n",
    "        \n",
    "    return local, online, greatpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: The default enrichment tables contain no associated genes for the input\n",
      "regions.You can set `download_by = 'tsv'` to download the complete\n",
      "table,but note only the top 500 regions can be retreived. See the\n",
      "following link:\n",
      "\n",
      "https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655401/Export#Export-GlobalExport\n",
      "\n",
      "R[write to console]: * check whether TxDb package 'TxDb.Hsapiens.UCSC.hg38.knownGene' is installed.\n",
      "\n",
      "R[write to console]: * gene ID type in the extended TSS is 'Entrez Gene ID'.\n",
      "\n",
      "R[write to console]: * restrict chromosomes to 'chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10,\n",
      "    chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX,\n",
      "    chrY, chrM'.\n",
      "\n",
      "R[write to console]: * 18208/29644 protein-coding genes left.\n",
      "\n",
      "R[write to console]: * update seqinfo to the selected chromosomes.\n",
      "\n",
      "R[write to console]: * TSS extension mode is 'basalPlusExt'.\n",
      "\n",
      "R[write to console]: * construct the basal domains by extending 5000bp to upstream and 1000bp to downsteram of TSS.\n",
      "\n",
      "R[write to console]: * calculate distances to neighbour regions.\n",
      "\n",
      "R[write to console]: * extend to both sides until reaching the neighbour genes or to the maximal extension (1000000bp).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "local,online,greatpy = rgreat_online_vs_local_vs_greatpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>binom_p_val</th>\n",
       "      <th>hyper_p_val</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>proteoglycan metabolic process</td>\n",
       "      <td>3.44454e-09</td>\n",
       "      <td>9.69768e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chondroitin sulfate proteoglycan metabolic pro...</td>\n",
       "      <td>6.60888e-09</td>\n",
       "      <td>1.25454e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>positive regulation of glycoprotein metabolic ...</td>\n",
       "      <td>8.50096e-08</td>\n",
       "      <td>5.22013e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>muscle cell migration</td>\n",
       "      <td>2.35739e-07</td>\n",
       "      <td>1.96183e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>replacement ossification</td>\n",
       "      <td>2.81099e-07</td>\n",
       "      <td>7.21346e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1830</th>\n",
       "      <td>inorganic ion transmembrane transport</td>\n",
       "      <td>9.95850e-01</td>\n",
       "      <td>9.78588e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1831</th>\n",
       "      <td>synapse organization</td>\n",
       "      <td>9.97051e-01</td>\n",
       "      <td>6.84269e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1832</th>\n",
       "      <td>neuron projection</td>\n",
       "      <td>9.98345e-01</td>\n",
       "      <td>7.85672e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1833</th>\n",
       "      <td>synapse</td>\n",
       "      <td>9.98970e-01</td>\n",
       "      <td>7.43814e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1834</th>\n",
       "      <td>synaptic signaling</td>\n",
       "      <td>9.99470e-01</td>\n",
       "      <td>9.39998e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1835 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   name  binom_p_val  \\\n",
       "0                        proteoglycan metabolic process  3.44454e-09   \n",
       "1     chondroitin sulfate proteoglycan metabolic pro...  6.60888e-09   \n",
       "2     positive regulation of glycoprotein metabolic ...  8.50096e-08   \n",
       "3                                 muscle cell migration  2.35739e-07   \n",
       "4                              replacement ossification  2.81099e-07   \n",
       "...                                                 ...          ...   \n",
       "1830              inorganic ion transmembrane transport  9.95850e-01   \n",
       "1831                               synapse organization  9.97051e-01   \n",
       "1832                                  neuron projection  9.98345e-01   \n",
       "1833                                            synapse  9.98970e-01   \n",
       "1834                                 synaptic signaling  9.99470e-01   \n",
       "\n",
       "      hyper_p_val  \n",
       "0     9.69768e-04  \n",
       "1     1.25454e-02  \n",
       "2     5.22013e-02  \n",
       "3     1.96183e-05  \n",
       "4     7.21346e-04  \n",
       "...           ...  \n",
       "1830  9.78588e-01  \n",
       "1831  6.84269e-01  \n",
       "1832  7.85672e-01  \n",
       "1833  7.43814e-01  \n",
       "1834  9.39998e-01  \n",
       "\n",
       "[1835 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>binom_p_val</th>\n",
       "      <th>hyper_p_val</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0017116</td>\n",
       "      <td>single-stranded dna-dependent atp-dependent dn...</td>\n",
       "      <td>3.68373e-11</td>\n",
       "      <td>4.36089e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0043141</td>\n",
       "      <td>atp-dependent 5'-3' dna helicase activity</td>\n",
       "      <td>5.41251e-11</td>\n",
       "      <td>7.16235e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0003680</td>\n",
       "      <td>at dna binding</td>\n",
       "      <td>8.83561e-11</td>\n",
       "      <td>2.49782e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0042975</td>\n",
       "      <td>peroxisome proliferator activated receptor bin...</td>\n",
       "      <td>1.10785e-10</td>\n",
       "      <td>9.02449e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0043139</td>\n",
       "      <td>5'-3' dna helicase activity</td>\n",
       "      <td>2.02135e-10</td>\n",
       "      <td>1.12120e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19104</th>\n",
       "      <td>GO:1990909</td>\n",
       "      <td>wnt signalosome</td>\n",
       "      <td>1.00000e+00</td>\n",
       "      <td>1.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19105</th>\n",
       "      <td>GO:1990913</td>\n",
       "      <td>sperm head plasma membrane</td>\n",
       "      <td>1.00000e+00</td>\n",
       "      <td>1.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19106</th>\n",
       "      <td>GO:1990917</td>\n",
       "      <td>ooplasm</td>\n",
       "      <td>1.00000e+00</td>\n",
       "      <td>1.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19107</th>\n",
       "      <td>GO:1990923</td>\n",
       "      <td>pet complex</td>\n",
       "      <td>1.00000e+00</td>\n",
       "      <td>1.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19108</th>\n",
       "      <td>GO:1990971</td>\n",
       "      <td>emilin complex</td>\n",
       "      <td>1.00000e+00</td>\n",
       "      <td>1.00000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19109 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id                                               name  \\\n",
       "0      GO:0017116  single-stranded dna-dependent atp-dependent dn...   \n",
       "1      GO:0043141          atp-dependent 5'-3' dna helicase activity   \n",
       "2      GO:0003680                                     at dna binding   \n",
       "3      GO:0042975  peroxisome proliferator activated receptor bin...   \n",
       "4      GO:0043139                        5'-3' dna helicase activity   \n",
       "...           ...                                                ...   \n",
       "19104  GO:1990909                                    wnt signalosome   \n",
       "19105  GO:1990913                         sperm head plasma membrane   \n",
       "19106  GO:1990917                                            ooplasm   \n",
       "19107  GO:1990923                                        pet complex   \n",
       "19108  GO:1990971                                     emilin complex   \n",
       "\n",
       "       binom_p_val  hyper_p_val  \n",
       "0      3.68373e-11  4.36089e-02  \n",
       "1      5.41251e-11  7.16235e-02  \n",
       "2      8.83561e-11  2.49782e-04  \n",
       "3      1.10785e-10  9.02449e-03  \n",
       "4      2.02135e-10  1.12120e-01  \n",
       "...            ...          ...  \n",
       "19104  1.00000e+00  1.00000e+00  \n",
       "19105  1.00000e+00  1.00000e+00  \n",
       "19106  1.00000e+00  1.00000e+00  \n",
       "19107  1.00000e+00  1.00000e+00  \n",
       "19108  1.00000e+00  1.00000e+00  \n",
       "\n",
       "[19109 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "online"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go_term</th>\n",
       "      <th>binom_p_value</th>\n",
       "      <th>hypergeom_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0031936</th>\n",
       "      <td>negative regulation of chromatin silencing</td>\n",
       "      <td>1.00612e-14</td>\n",
       "      <td>1.40000e-307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0017116</th>\n",
       "      <td>single-stranded dna-dependent atp-dependent dn...</td>\n",
       "      <td>4.80098e-12</td>\n",
       "      <td>3.00000e-308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042974</th>\n",
       "      <td>retinoic acid receptor binding</td>\n",
       "      <td>1.25139e-11</td>\n",
       "      <td>1.50000e-307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0003680</th>\n",
       "      <td>at dna binding</td>\n",
       "      <td>4.80863e-11</td>\n",
       "      <td>1.50000e-307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0043141</th>\n",
       "      <td>atp-dependent 5'-3' dna helicase activity</td>\n",
       "      <td>5.19613e-11</td>\n",
       "      <td>8.00000e-308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005886</th>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>9.94321e-01</td>\n",
       "      <td>2.22000e-306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0043025</th>\n",
       "      <td>neuronal cell body</td>\n",
       "      <td>9.95491e-01</td>\n",
       "      <td>2.76000e-306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0007420</th>\n",
       "      <td>brain development</td>\n",
       "      <td>9.97816e-01</td>\n",
       "      <td>2.13000e-306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0030424</th>\n",
       "      <td>axon</td>\n",
       "      <td>9.99016e-01</td>\n",
       "      <td>2.77000e-306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0030425</th>\n",
       "      <td>dendrite</td>\n",
       "      <td>9.99892e-01</td>\n",
       "      <td>2.77000e-306</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2287 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      go_term binom_p_value  \\\n",
       "GO:0031936         negative regulation of chromatin silencing   1.00612e-14   \n",
       "GO:0017116  single-stranded dna-dependent atp-dependent dn...   4.80098e-12   \n",
       "GO:0042974                     retinoic acid receptor binding   1.25139e-11   \n",
       "GO:0003680                                     at dna binding   4.80863e-11   \n",
       "GO:0043141          atp-dependent 5'-3' dna helicase activity   5.19613e-11   \n",
       "...                                                       ...           ...   \n",
       "GO:0005886                                    plasma membrane   9.94321e-01   \n",
       "GO:0043025                                 neuronal cell body   9.95491e-01   \n",
       "GO:0007420                                  brain development   9.97816e-01   \n",
       "GO:0030424                                               axon   9.99016e-01   \n",
       "GO:0030425                                           dendrite   9.99892e-01   \n",
       "\n",
       "           hypergeom_p_value  \n",
       "GO:0031936      1.40000e-307  \n",
       "GO:0017116      3.00000e-308  \n",
       "GO:0042974      1.50000e-307  \n",
       "GO:0003680      1.50000e-307  \n",
       "GO:0043141      8.00000e-308  \n",
       "...                      ...  \n",
       "GO:0005886      2.22000e-306  \n",
       "GO:0043025      2.76000e-306  \n",
       "GO:0007420      2.13000e-306  \n",
       "GO:0030424      2.77000e-306  \n",
       "GO:0030425      2.77000e-306  \n",
       "\n",
       "[2287 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greatpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "local.to_csv(\"../../../local.csv\",sep=\"\\t\",header=True,index=False)\n",
    "online.to_csv(\"../../../online.csv\",sep=\"\\t\",header=True,index=False)\n",
    "greatpy.to_csv(\"../../../greatpy.csv\",sep=\"\\t\",header=True,index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "sns.set_theme()\n",
    "\n",
    "# Load the brain networks example dataset\n",
    "df = sns.load_dataset(\"brain_networks\", header=[0, 1, 2], index_col=0)\n",
    "\n",
    "# Select a subset of the networks\n",
    "used_networks = [1, 5, 6, 7, 8, 12, 13, 17]\n",
    "used_columns = (df.columns.get_level_values(\"network\")\n",
    "                          .astype(int)\n",
    "                          .isin(used_networks))\n",
    "df = df.loc[:, used_columns]\n",
    "\n",
    "# Create a categorical palette to identify the networks\n",
    "network_pal = sns.husl_palette(8, s=.45)\n",
    "network_lut = dict(zip(map(str, used_networks), network_pal))\n",
    "\n",
    "# Convert the palette to vectors that will be drawn on the side of the matrix\n",
    "networks = df.columns.get_level_values(\"network\")\n",
    "network_colors = pd.Series(networks, index=df.columns).map(network_lut)\n",
    "\n",
    "# Draw the full plot\n",
    "g = sns.clustermap(df.corr(), center=0, cmap=\"vlag\",\n",
    "                   row_colors=network_colors, col_colors=network_colors,\n",
    "                   dendrogram_ratio=(.1, .2),\n",
    "                   cbar_pos=(.02, .32, .03, .2),\n",
    "                   linewidths=.75, figsize=(12, 13))\n",
    "\n",
    "g.ax_row_dendrogram.remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns; sns.set_theme(color_codes=True)\n",
    "iris = sns.load_dataset(\"iris\")\n",
    "species = iris.pop(\"species\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'setosa': 'r', 'versicolor': 'b', 'virginica': 'g'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lut = dict(zip(species.unique(), \"rbg\"))\n",
    "lut\n",
    "# row_colors = species.map(lut)\n",
    "# g = sns.clustermap(iris, row_colors=row_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# id = online.pop(\"id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import random\n",
    "# col = {i : \"#\"+''.join([random.choice('0123456789ABCDEF') for j in range(6)]) for i in id.unique()}\n",
    "# row_colors = id.map(col)\n",
    "# sns.clustermap(online[[\"binom_p_val\",\"hyper_p_val\"]], row_colors=row_colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see the correlation between greatpy and local analysis of rGreat is not good."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'preprocess' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb Cell 20\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell://wsl%2Bubuntu/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb#ch0000011vscode-remote?line=0'>1</a>\u001b[0m preprocess\n",
      "\u001b[0;31mNameError\u001b[0m: name 'preprocess' is not defined"
     ]
    }
   ],
   "source": [
    "preprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, the shape reduction is very high so It would to say : the GO term pick by the two function is diffrent."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('gpy')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "610c73e5f1997caa668091b43a573fbafba1e2b96e55fbc05d56c58d21b90949"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
