{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rGreat local vs greatpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook allows you to compare results between local rGreat and enrichment function from greatpy. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "r-base == 3.6.1\n",
    "\n",
    "python version == 3.8 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects import r as r\n",
    "pandas2ri.activate()\n",
    "import pandas as pd\n",
    "\n",
    "import os \n",
    "import matplotlib.pyplot as plt\n",
    "import greatpy as great \n",
    "from scipy.stats import pearsonr\n",
    "import re\n",
    "import time \n",
    "import seaborn as sns \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.5.2'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpy2.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "importr('rGREAT')\n",
    "ranges = importr('GenomicRanges')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rgreat_online_vs_local_vs_greatpy(): \n",
    "    stat_df = {\n",
    "        \"name\" : [],\n",
    "        \"pearson_binom\" : [], \n",
    "        \"pearson_hypergeom\" : []\n",
    "    }\n",
    "    pp = {\n",
    "        \"name\" : [],\n",
    "        \"before_pp_greatpy_size\" : [], \n",
    "        \"before_pp_local_size\" : [],\n",
    "        \"final_size\" : [],\n",
    "        \"%_of_diffrent_GO_term\" : [],\n",
    "    }\n",
    "    test_file = [\"01_random.bed\",\"06_height_snps_hg38.bed\",\"07_height_snps_hg19.bed\",\"10_MAX.bed\"]\n",
    "    # test_file = [\"06_height_snps_hg38.bed\",\"10_MAX.bed\"]\n",
    "\n",
    "    # test_file = [\"06_height_snps_hg38.bed\"]\n",
    "\n",
    "    ann = pd.read_csv(\"../../data/human/ontologies.csv\",sep=';',comment='#',header=0,)\n",
    "    ann[\"name\"] = ann['name'].str.lower()\n",
    "\n",
    "    for name in test_file : \n",
    "        # find the assembly \n",
    "        if re.match(\".*hg19.*\",name) != None : \n",
    "            assembly = \"hg19\"\n",
    "        else : \n",
    "            assembly = \"hg38\"\n",
    "\n",
    "        # online test \n",
    "        res_online = rpy2.robjects.r['submitGreatJob'](f\"../../data/tests/test_data/input/{name}\",species=f\"{assembly}\",help=False)\n",
    "        res_online = rpy2.robjects.r['getEnrichmentTables'](res_online)\n",
    "\n",
    "        time.sleep(30)\n",
    "\n",
    "        # local test\n",
    "            # proprocessing : make a Grange frame \n",
    "        df = r[\"read.csv\"](f\"../../data/tests/test_data/input/{name}\",sep='\\t')\n",
    "        seqname = rpy2.robjects.StrVector([\"seqnames\", \"seqname\",\"chromosome\", \"X.Chr\",\"chr\", \"chromosome_name\",\"seqid\"])\n",
    "        end = rpy2.robjects.StrVector(['end', 'stop'])\n",
    "        df = ranges.makeGRangesFromDataFrame(df,seqnames_field=seqname)\n",
    "\n",
    "            # great calculation\n",
    "        local = rpy2.robjects.r['great'](df, \"GREAT:C5\", f\"txdb:{assembly}\",verbose=False)\n",
    "        local = rpy2.robjects.r['getEnrichmentTables'](local)\n",
    "\n",
    "        time.sleep(30)\n",
    "\n",
    "        # greatpy calculation\n",
    "        greatpy = great.tl.GREAT.enrichment(\n",
    "            test_file=f\"../../data/tests/test_data/input/{name}\",\n",
    "            regdom_file=f\"../../data/human/{assembly}/regulatory_domain.bed\",\n",
    "            chr_size_file=f\"../../data/human/{assembly}/chr_size.bed\",\n",
    "            annotation_file=f\"../../data/human/ontologies.csv\",\n",
    "            binom=True,\n",
    "            hypergeom=True,\n",
    "            )\n",
    "        time.sleep(30)\n",
    "\n",
    "        # create each dataframe\n",
    "            # online\n",
    "        name_online = [cdc.lower() for cdc in list(res_online.rx2(\"GO Molecular Function\").rx2(\"name\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"name\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"name\"))]\n",
    "        online = pd.DataFrame({\n",
    "            \"id\" : list(res_online.rx2(\"GO Molecular Function\").rx2(\"ID\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"ID\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"ID\")),\n",
    "            \"name\" : name_online,\n",
    "            \"binom_p_val\" : list(res_online.rx2(\"GO Molecular Function\").rx2(\"Binom_Raw_PValue\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"Binom_Raw_PValue\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"Binom_Raw_PValue\")),\n",
    "            \"hyper_p_val\" : list(res_online.rx2(\"GO Molecular Function\").rx2(\"Hyper_Raw_PValue\"))+list(res_online.rx2(\"GO Biological Process\").rx2(\"Hyper_Raw_PValue\"))+list(res_online.rx2(\"GO Cellular Component\").rx2(\"Hyper_Raw_PValue\"))\n",
    "            })\n",
    "\n",
    "            # local\n",
    "        name_local = list(local.rx2(\"id\")) \n",
    "        name_local = [\" \".join(cdc.lower().split(\"_\")[1:]) for cdc in list(local.rx2(\"id\"))]\n",
    "        local = pd.DataFrame({\n",
    "            \"name\": name_local,\n",
    "            \"binom_p_val\" : list(local.rx2(\"p_value\")),\n",
    "            \"hyper_p_val\" : list(local.rx2(\"p_value_hyper\"))\n",
    "            })\n",
    "\n",
    "            # greatpy\n",
    "        greatpy[\"go_term\"] = greatpy[\"go_term\"].str.lower()\n",
    "\n",
    "        # correlation between online and greatpy \n",
    "        o_great = online[online[\"id\"].isin(list(greatpy.index))]\n",
    "        great_o = greatpy[greatpy.index.isin(list(online[\"id\"]))]\n",
    "\n",
    "        tot = pd.DataFrame()\n",
    "        res_bin=[]\n",
    "        res_hyp=[]\n",
    "        cols_to_supr = []\n",
    "        for i in range(len(o_great)) : \n",
    "            curr_o = o_great.iloc[i]\n",
    "            if greatpy.loc[greatpy.index == curr_o[\"id\"]].shape[0] > 0 : \n",
    "                res_bin.append( float(greatpy.loc[greatpy.index == curr_o[\"id\"]][\"binom_p_value\"]) )\n",
    "                res_hyp.append( float( greatpy.loc[greatpy.index == curr_o[\"id\"]][\"hypergeom_p_value\"] ))\n",
    "            else : \n",
    "                cols_to_supr.append(i)\n",
    "        \n",
    "        tot[\"id\"] = o_great[\"id\"]\n",
    "        tot[\"binom_p_value_online\"] = -np.log(list(o_great[\"binom_p_val\"]))\n",
    "        tot[\"hyper_p_value_online\"] = -np.log(list(o_great[\"hyper_p_val\"]))\n",
    "        tot[\"binom_p_value_greatpy\"] = -np.log(res_bin)\n",
    "        tot[\"hypergeom_p_value_greatpy\"] = -np.log(res_hyp)\n",
    "\n",
    "        # add GO id to local \n",
    "        go = []\n",
    "        for i in range(local.shape[0]) : \n",
    "            curr = local.iloc[i]\n",
    "\n",
    "            if ann.loc[ann[\"term\"].isin([curr[\"name\"]])].shape[0]>0 : \n",
    "                go.append(ann.loc[ann[\"term\"].isin([curr[\"name\"]])].iloc[0][\"id\"])\n",
    "            else :\n",
    "                go.append(\"\")\n",
    "        local[\"id\"] = go \n",
    "        local = local.loc[local[\"id\"]!=\"\"]\n",
    "\n",
    "        # Correlation between online and local \n",
    "        o_loc = online[online[\"id\"].isin(local[\"id\"])]\n",
    "        loc_o = local[local[\"id\"].isin(online[\"id\"])]\n",
    "        tot_2 = pd.DataFrame()\n",
    "\n",
    "        res_bin=[]\n",
    "        res_hyp=[]\n",
    "        cols_to_supr = []\n",
    "        for i in range(len(o_loc)) : \n",
    "            curr_o = o_loc.iloc[i]\n",
    "            if loc_o.loc[loc_o[\"id\"] == curr_o[\"id\"]].shape[0] > 0 : \n",
    "                res_bin.append(float(loc_o.loc[loc_o[\"id\"] == curr_o[\"id\"]].iloc[0][\"binom_p_val\"]) )\n",
    "                res_hyp.append(float(loc_o.loc[loc_o[\"id\"] == curr_o[\"id\"]].iloc[0][\"hyper_p_val\"] ))\n",
    "            else : \n",
    "                cols_to_supr.append(i)\n",
    "        tot_2[\"id\"] = o_loc[\"id\"]\n",
    "        tot_2[\"binom_p_value_online\"] = -np.log(list(o_loc[\"binom_p_val\"]))\n",
    "        tot_2[\"hyper_p_value_online\"] = -np.log(list(o_loc[\"hyper_p_val\"]))\n",
    "        tot_2[\"binom_p_value_local\"] = -np.log(res_bin)\n",
    "        tot_2[\"hypergeom_p_value_local\"] = -np.log(res_hyp)\n",
    "\n",
    "        # Correlation between greatpy and local \n",
    "        g_loc = greatpy[greatpy.index.isin(local[\"id\"])]\n",
    "        loc_g = local[local[\"id\"].isin(greatpy.index)]\n",
    "        tot_3 = pd.DataFrame()\n",
    "\n",
    "        res_bin=[]\n",
    "        res_hyp=[]\n",
    "        cols_to_supr = []\n",
    "        g_loc = g_loc.reset_index().rename(columns={\"index\":\"id\"})\n",
    "        for i in range(g_loc.shape[0]) :  \n",
    "            curr_o = g_loc.iloc[i]\n",
    "            if loc_g.loc[loc_g[\"id\"] == curr_o[\"id\"]].shape[0] > 0 : \n",
    "                res_bin.append(float(loc_g.loc[loc_g[\"id\"] == curr_o[\"id\"]].iloc[0][\"binom_p_val\"]) )\n",
    "                res_hyp.append(float(loc_g.loc[loc_g[\"id\"] == curr_o[\"id\"]].iloc[0][\"hyper_p_val\"] ))\n",
    "            else : \n",
    "                cols_to_supr.append(i)\n",
    "        tot_3[\"id\"] = g_loc[\"id\"]\n",
    "        tot_3[\"binom_p_value_greatpy\"] = -np.log(list(g_loc[\"binom_p_value\"]))\n",
    "        tot_3[\"hyper_p_value_greatpy\"] = -np.log(list(g_loc[\"hypergeom_p_value\"]))\n",
    "        tot_3[\"binom_p_value_local\"] = -np.log(res_bin)\n",
    "        tot_3[\"hypergeom_p_value_local\"] = -np.log(res_hyp)\n",
    "\n",
    "        tot = tot.replace(np.inf,np.nan)\n",
    "        tot_2 = tot_2.replace(np.inf,np.nan)\n",
    "        tot_3 = tot_3.replace(np.inf,np.nan)\n",
    "        \n",
    "        tot = tot.dropna()\n",
    "        tot_2 = tot_2.dropna()\n",
    "        tot_3 = tot_3.dropna()\n",
    "\n",
    "        # create the table \n",
    "        res = {\"name1\":[],\"name2\":[],\"value\":[]}\n",
    "\n",
    "        res[\"name1\"].append(\"greatpy\")\n",
    "        res[\"name2\"].append(\"greatpy\")\n",
    "        res[\"value\"].append(pearsonr(tot[\"binom_p_value_greatpy\"],tot[\"binom_p_value_greatpy\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"online\")\n",
    "        res[\"name2\"].append(\"greatpy\")\n",
    "        res[\"value\"].append(pearsonr(tot[\"binom_p_value_online\"],tot[\"binom_p_value_greatpy\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"greatpy\")\n",
    "        res[\"name2\"].append(\"online\")\n",
    "        res[\"value\"].append(pearsonr(tot[\"binom_p_value_greatpy\"],tot[\"binom_p_value_online\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"online\")\n",
    "        res[\"name2\"].append(\"local\")\n",
    "        res[\"value\"].append(pearsonr(tot_2[\"binom_p_value_online\"],tot_2[\"binom_p_value_local\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"local\")\n",
    "        res[\"name2\"].append(\"online\")\n",
    "        res[\"value\"].append(pearsonr(tot_2[\"binom_p_value_online\"],tot_2[\"binom_p_value_local\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"online\")\n",
    "        res[\"name2\"].append(\"online\")\n",
    "        res[\"value\"].append(pearsonr(tot[\"binom_p_value_online\"],tot[\"binom_p_value_online\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"local\")\n",
    "        res[\"name2\"].append(\"greatpy\")\n",
    "        res[\"value\"].append(pearsonr(tot_3[\"binom_p_value_greatpy\"],tot_3[\"binom_p_value_local\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"greatpy\")\n",
    "        res[\"name2\"].append(\"local\")\n",
    "        res[\"value\"].append(pearsonr(tot_3[\"binom_p_value_greatpy\"],tot_3[\"binom_p_value_local\"])[0])\n",
    "\n",
    "        res[\"name1\"].append(\"local\")\n",
    "        res[\"name2\"].append(\"local\")\n",
    "        res[\"value\"].append(pearsonr(tot_2[\"binom_p_value_local\"],tot_2[\"binom_p_value_local\"])[0])\n",
    "\n",
    "\n",
    "        test = pd.DataFrame(res)\n",
    "        test = test.pivot(\"name1\",\"name2\",\"value\")\n",
    "        g = sns.clustermap(test,figsize=(5,5),annot = True)\n",
    "        g.ax_heatmap.set_title(f\"{name}\")\n",
    "        \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: The default enrichment tables contain no associated genes for the input\n",
      "regions.You can set `download_by = 'tsv'` to download the complete\n",
      "table,but note only the top 500 regions can be retreived. See the\n",
      "following link:\n",
      "\n",
      "https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655401/Export#Export-GlobalExport\n",
      "\n",
      "R[write to console]: * check whether TxDb package 'TxDb.Hsapiens.UCSC.hg38.knownGene' is installed.\n",
      "\n",
      "R[write to console]: * gene ID type in the extended TSS is 'Entrez Gene ID'.\n",
      "\n",
      "R[write to console]: * restrict chromosomes to 'chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10,\n",
      "    chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX,\n",
      "    chrY, chrM'.\n",
      "\n",
      "R[write to console]: * 18208/29644 protein-coding genes left.\n",
      "\n",
      "R[write to console]: * update seqinfo to the selected chromosomes.\n",
      "\n",
      "R[write to console]: * TSS extension mode is 'basalPlusExt'.\n",
      "\n",
      "R[write to console]: * construct the basal domains by extending 5000bp to upstream and 1000bp to downsteram of TSS.\n",
      "\n",
      "R[write to console]: * calculate distances to neighbour regions.\n",
      "\n",
      "R[write to console]: * extend to both sides until reaching the neighbour genes or to the maximal extension (1000000bp).\n",
      "\n",
      "R[write to console]: \n",
      "\n"
     ]
    },
    {
     "ename": "RRuntimeError",
     "evalue": "Error in (function (classes, fdef, mtable)  : \n  unable to find an inherited method for function ‘dims’ for signature ‘\"DFrame\"’\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[1;32m/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb Cell 9\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell://wsl%2Bubuntu/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb#ch0000008vscode-remote?line=0'>1</a>\u001b[0m local,online,greatpy \u001b[39m=\u001b[39m rgreat_online_vs_local_vs_greatpy()\n",
      "\u001b[1;32m/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb Cell 9\u001b[0m in \u001b[0;36mrgreat_online_vs_local_vs_greatpy\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb#ch0000008vscode-remote?line=36'>37</a>\u001b[0m df \u001b[39m=\u001b[39m ranges\u001b[39m.\u001b[39mmakeGRangesFromDataFrame(df,seqnames_field\u001b[39m=\u001b[39mseqname)\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb#ch0000008vscode-remote?line=38'>39</a>\u001b[0m     \u001b[39m# great calculation\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell://wsl%2Bubuntu/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb#ch0000008vscode-remote?line=39'>40</a>\u001b[0m local \u001b[39m=\u001b[39m rpy2\u001b[39m.\u001b[39;49mrobjects\u001b[39m.\u001b[39;49mr[\u001b[39m'\u001b[39;49m\u001b[39mgreat\u001b[39;49m\u001b[39m'\u001b[39;49m](df, \u001b[39m\"\u001b[39;49m\u001b[39mmsigdb:C5\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mf\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mtxdb:\u001b[39;49m\u001b[39m{\u001b[39;49;00massembly\u001b[39m}\u001b[39;49;00m\u001b[39m\"\u001b[39;49m,verbose\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb#ch0000008vscode-remote?line=40'>41</a>\u001b[0m local \u001b[39m=\u001b[39m rpy2\u001b[39m.\u001b[39mrobjects\u001b[39m.\u001b[39mr[\u001b[39m'\u001b[39m\u001b[39mgetEnrichmentTables\u001b[39m\u001b[39m'\u001b[39m](local)\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/tom/Stage/greatpy/notebooks/test/06_rgreat_online_vs_local_vs_greatpy.ipynb#ch0000008vscode-remote?line=42'>43</a>\u001b[0m time\u001b[39m.\u001b[39msleep(\u001b[39m30\u001b[39m)\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/rpy2/robjects/functions.py:202\u001b[0m, in \u001b[0;36mSignatureTranslatedFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    200\u001b[0m         v \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(k)\n\u001b[1;32m    201\u001b[0m         kwargs[r_k] \u001b[39m=\u001b[39m v\n\u001b[0;32m--> 202\u001b[0m \u001b[39mreturn\u001b[39;00m (\u001b[39msuper\u001b[39;49m(SignatureTranslatedFunction, \u001b[39mself\u001b[39;49m)\n\u001b[1;32m    203\u001b[0m         \u001b[39m.\u001b[39;49m\u001b[39m__call__\u001b[39;49m(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs))\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/rpy2/robjects/functions.py:125\u001b[0m, in \u001b[0;36mFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    123\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    124\u001b[0m         new_kwargs[k] \u001b[39m=\u001b[39m cv\u001b[39m.\u001b[39mpy2rpy(v)\n\u001b[0;32m--> 125\u001b[0m res \u001b[39m=\u001b[39m \u001b[39msuper\u001b[39;49m(Function, \u001b[39mself\u001b[39;49m)\u001b[39m.\u001b[39;49m\u001b[39m__call__\u001b[39;49m(\u001b[39m*\u001b[39;49mnew_args, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mnew_kwargs)\n\u001b[1;32m    126\u001b[0m res \u001b[39m=\u001b[39m cv\u001b[39m.\u001b[39mrpy2py(res)\n\u001b[1;32m    127\u001b[0m \u001b[39mreturn\u001b[39;00m res\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/rpy2/rinterface_lib/conversion.py:45\u001b[0m, in \u001b[0;36m_cdata_res_to_rinterface.<locals>._\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_\u001b[39m(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[0;32m---> 45\u001b[0m     cdata \u001b[39m=\u001b[39m function(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     46\u001b[0m     \u001b[39m# TODO: test cdata is of the expected CType\u001b[39;00m\n\u001b[1;32m     47\u001b[0m     \u001b[39mreturn\u001b[39;00m _cdata_to_rinterface(cdata)\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/rpy2/rinterface.py:813\u001b[0m, in \u001b[0;36mSexpClosure.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    806\u001b[0m     res \u001b[39m=\u001b[39m rmemory\u001b[39m.\u001b[39mprotect(\n\u001b[1;32m    807\u001b[0m         openrlib\u001b[39m.\u001b[39mrlib\u001b[39m.\u001b[39mR_tryEval(\n\u001b[1;32m    808\u001b[0m             call_r,\n\u001b[1;32m    809\u001b[0m             call_context\u001b[39m.\u001b[39m__sexp__\u001b[39m.\u001b[39m_cdata,\n\u001b[1;32m    810\u001b[0m             error_occured)\n\u001b[1;32m    811\u001b[0m     )\n\u001b[1;32m    812\u001b[0m     \u001b[39mif\u001b[39;00m error_occured[\u001b[39m0\u001b[39m]:\n\u001b[0;32m--> 813\u001b[0m         \u001b[39mraise\u001b[39;00m embedded\u001b[39m.\u001b[39mRRuntimeError(_rinterface\u001b[39m.\u001b[39m_geterrmessage())\n\u001b[1;32m    814\u001b[0m \u001b[39mreturn\u001b[39;00m res\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in (function (classes, fdef, mtable)  : \n  unable to find an inherited method for function ‘dims’ for signature ‘\"DFrame\"’\n"
     ]
    }
   ],
   "source": [
    "local,online,greatpy = rgreat_online_vs_local_vs_greatpy()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
