{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext line_profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from math import lgamma, log, exp,fabs,inf\n",
    "from scipy.special import comb\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import dask.dataframe as dd \n",
    "import numpy as np \n",
    "\n",
    "pd.options.display.float_format = '{:12.5e}'.format\n",
    "\n",
    "def get_association(test,regdom): \n",
    "    \"\"\"\n",
    "    Function allowing from a file of genomic regions from CHIPseq \n",
    "    and a file of genomic regulatory domains to determine the names \n",
    "    of genes associated with at least one genomic region \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    test : pd.dataFrame\n",
    "        df of the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    res : list\n",
    "        list of gene associated with at least with one test peak\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    test = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1052028],\n",
    "    ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']})\n",
    "\n",
    "    >>> get_association(test,regdom)        \n",
    "        ['RNF223']\n",
    "    \n",
    "    \"\"\"\n",
    "    res = []\n",
    "    for i in range(test.shape[0]):\n",
    "        currTest = test.iloc[i]\n",
    "        regdom_curr_test = regdom.loc[(regdom[\"Chr\"] == currTest[\"Chr\"])].sort_values(\"Chr_Start\")\n",
    "        regdom_curr_test = regdom_curr_test.loc[((regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_End\"])) | ((regdom_curr_test[\"Chr_Start\"] >= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] <= currTest[\"Chr_End\"])) | ((regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_End\"]))] \n",
    "        res = res + list(regdom_curr_test[\"Name\"])\n",
    "    return list(dict.fromkeys(res))\n",
    "\n",
    "def len_regdom(regdom:pd.DataFrame): \n",
    "    \"\"\"\n",
    "    Function to calculate for each gene name from regdom the\n",
    "     size of the regulatory region for this gene in the genome \n",
    "\n",
    "    Parameters\n",
    "    ----------    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dict\n",
    "        dictionary in which each key corresponds to a gene name \n",
    "        from regdom and the value is the size of the regulatory \n",
    "        region for that gene\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']}))\n",
    "\n",
    "    >>> len_regdom(regdom)\n",
    "        {'RNF223': 80097, 'C1orf159': 52710}\n",
    "\n",
    "    \"\"\"\n",
    "    test = regdom[\"Chr_End\"]-regdom[\"Chr_Start\"]\n",
    "    return pd.DataFrame({\"len\":list(test)},index=regdom[\"Name\"]).to_dict()[\"len\"]\n",
    "\n",
    "def number_of_hit(test,regdom): \n",
    "    \"\"\" \n",
    "    Function to calculate the number of hits from several \n",
    "    genomic regions and the file describing the regulatory regions\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    test : pd.dataFrame\n",
    "        df of the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    nb : int\n",
    "        number of hit \n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    test = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1052028],\n",
    "    ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']})\n",
    "\n",
    "    >>> number_of_hit(test,regdom)        \n",
    "        1\n",
    "    \n",
    "    \"\"\"\n",
    "    nb=0\n",
    "    regdom = regdom[[\"Chr\",\"Chr_Start\",\"Chr_End\"]]\n",
    "    for i in range(test.shape[0]): \n",
    "        chrom = test.iat[i,0]\n",
    "        start = test.iat[i,1]\n",
    "        end = test.iat[i,2]\n",
    "        regdom_np = regdom[\"Chr\"].to_numpy()\n",
    "        reg_start = regdom[\"Chr_Start\"].to_numpy()\n",
    "        reg_end = regdom[\"Chr_End\"].to_numpy()\n",
    "        Chr_reduce = np.where(regdom_np == chrom)\n",
    "        reg_start = np.take(reg_start,Chr_reduce,axis=0)[0]\n",
    "        reg_end = np.take(reg_end,Chr_reduce,axis=0)[0]\n",
    "\n",
    "        if any((reg_start <= start) & (reg_end >= end)):  \n",
    "            nb += 1\n",
    "    return nb\n",
    "\n",
    "def betacf(a,b,x): \n",
    "    \"\"\" Used by betai: Evaluates continued fraction for incomplete beta function \"\"\"\n",
    "    maxit = 10000\n",
    "    eps = 3.0e-7 \n",
    "    fpmin = 1.0e-30\n",
    "    qab = a+b\n",
    "    qap = a+1\n",
    "    qam = a-1\n",
    "    c = 1\n",
    "    d = 1-qab*x/qap\n",
    "    if fabs(d) < fpmin :\n",
    "        d = fpmin\n",
    "    d = 1/d \n",
    "    h = d\n",
    "    for m in range(1,maxit+1): \n",
    "        m2 = 2*m\n",
    "        aa = m*(b-m)*x/((qam+m2)*(a+m2))\n",
    "        d = 1.0+aa*d\n",
    "        if (fabs(d) < fpmin) : \n",
    "            d = fpmin\n",
    "        c = 1.0+aa/c\n",
    "        if (fabs(c) < fpmin):\n",
    "            c = fpmin\n",
    "        d = 1.0/d\n",
    "        h *= d*c\n",
    "        aa = -(a+m)*(qab+m)*x/((a+m2)*(qap+m2))\n",
    "        d = 1.0+aa*d  \n",
    "        if (fabs(d) < fpmin):\n",
    "            d = fpmin\n",
    "        c = 1.0+aa/c\n",
    "        if (fabs(c) < fpmin):\n",
    "            c = fpmin\n",
    "        d = 1.0/d\n",
    "        dell = d*c\n",
    "        h *= dell\n",
    "        if (fabs(dell-1.0) < eps):\n",
    "            break\n",
    "    if (m > maxit):\n",
    "        print(\"a or b too big, or MAXIT too small in betacf\")\n",
    "        return False\n",
    "    return h\n",
    "\n",
    "def betai(a,b,x):\n",
    "    \"\"\"Returns the incomplete beta function Ix(a, b).\"\"\"\n",
    "    if x < 0 or x > 1 : \n",
    "        return False\n",
    "    if x == 0 or x == 1 : \n",
    "        bt = 0.0\n",
    "    else : \n",
    "        bt = exp(lgamma(a+b)-lgamma(a)-lgamma(b)+a*log(x)+b*log(1.0-x))\n",
    "    if x < (a+1)/(a+b+2) : \n",
    "        return bt*betacf(a,b,x)/a\n",
    "    return 1-bt*betacf(b,a,1-x)/b\n",
    "\n",
    "def get_binom_pval(n:int,k:int,p:float) -> float:\n",
    "    \"\"\"\n",
    "    This function allows to calculate the binomial probability \n",
    "    of obtaining k in a set of size n and whose probability is p \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n : int\n",
    "        Number of genomic region in the test set \n",
    "    k : int \n",
    "        Number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "    p : float\n",
    "        Percentage of genome annotated\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        binomial probability\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> get_binom_pval(100,2,0.2)\n",
    "        0.9999999947037065\n",
    "    \n",
    "    \"\"\"\n",
    "    if k == 0 : return 1\n",
    "    else : return betai(k,n-k+1,p)\n",
    "\n",
    "def hypergeom_pmf(N, K, n, k):\n",
    "    \"\"\"\n",
    "    Function to calculate the probability mass function for hypergeometric distribution\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    N : int\n",
    "        Total number of gene in the genome\n",
    "    K : int \n",
    "        Number of genes in the genome with annotation\n",
    "    n : int\n",
    "        Number of gene in the test set\n",
    "    k : int\n",
    "        Number of genes in the test gene set with annotation\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        proability mass function\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> hypergeom_pmf(100,10,30,1)\n",
    "        0.11270773995748315\n",
    "    \n",
    "    \"\"\"\n",
    "    Achoosex = comb(K,k) if comb(K,k) != inf else 1e-308\n",
    "    NAchoosenx = comb(N-K, n-k) if comb(N-K, n-k) != inf else 1e-308\n",
    "    Nchoosen = comb(N,n) if comb(N,n) != inf else 1e-308\n",
    "    return ((Achoosex)*NAchoosenx)/Nchoosen if Nchoosen > 1e-308 and (Achoosex)*NAchoosenx != 0.0 else 1e-308\n",
    "\n",
    "def hypergeom_cdf(N, K, n, k):\n",
    "    \"\"\"\n",
    "    Function to calculate the cumulative density funtion for hypergeometric distribution\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    N : int\n",
    "        Total number of gene in the genome\n",
    "    K : int \n",
    "        Number of genes in the genome with annotation\n",
    "    n : int\n",
    "        Number of gene in the test set\n",
    "    k : int\n",
    "        Number of genes in the test gene set with annotation\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        Cumulative density function\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> hypergeom_cdf(100,10,30,1)\n",
    "        0.9770827595419788\n",
    "    \n",
    "    \"\"\"\n",
    "    return np.sum([hypergeom_pmf(N, K, n, x) for x in range(k,min(K,n)+1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def loader(test_data:None or str or pd.DataFrame,regdom_file:None or str or pd.DataFrame,chr_size_file:None or str or pd.DataFrame,annotation_file:None or str or pd.DataFrame):\n",
    "    if type(regdom_file) == str:\n",
    "        regdom = pd.read_csv(regdom_file,sep=\"\\t\",comment=\"#\",\n",
    "                    names=[\"Chr\", \"Chr_Start\", \"Chr_End\",\"Name\",\"tss\",\"Strand\"],dtype={\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\",\"Name\":\"object\",\"tss\":\"int64\",\"Strand\":\"object\"})\n",
    "    elif type(regdom_file) == pd.DataFrame:\n",
    "        regdom = regdom_file.iloc[:,:6]\n",
    "        colname = list(regdom.columns)\n",
    "        try : \n",
    "            regdom = regdom.rename(columns={colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\",colname[3]:\"Name\",colname[4]:\"tss\",colname[5]:\"Strand\"})\n",
    "        except :\n",
    "            print(\"Error in the format of the regdom file\")\n",
    "            print(\"The regdom file must have the following columns : Chr, Chr_Start, Chr_End, Name, tss, Strand\")\n",
    "            return False \n",
    "    else : \n",
    "        regdom = regdom_file\n",
    "\n",
    "    if type(test_data) == str : \n",
    "        test_data = pd.read_csv(test_data,sep=\"\\t\",comment=\"#\",usecols=[0,1,2],\n",
    "                        names=[\"Chr\", \"Chr_Start\", \"Chr_End\"],dtype={\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\"})\n",
    "    elif type(test_data) == pd.DataFrame : \n",
    "        test_data = test_data.iloc[:,:3]\n",
    "        colname = list(test_data.columns)\n",
    "        try : \n",
    "            test_data = test_data.rename(columns={colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\"})\n",
    "        except : \n",
    "            print(\"Error in test dataframe, please check your input\")\n",
    "            print(\"Columns should be : chr...(type object), start(type int), end(type int)\")\n",
    "            return False\n",
    "    else :\n",
    "        pass\n",
    "\n",
    "    \n",
    "    if type(chr_size_file) == str :\n",
    "        size = pd.read_csv(chr_size_file,sep=\"\\t\",comment=\"#\",\n",
    "                        names=[\"Chrom\",\"Size\"],dtype={\"Chrom\":\"object\", \"Size\":\"int64\"})\n",
    "    elif type(chr_size_file) == pd.DataFrame :\n",
    "        size = chr_size_file.iloc[:,:2]\n",
    "        colname = list(size.columns)\n",
    "        try : \n",
    "            size = size.rename(columns={colname[0]:\"Chrom\",colname[1]:\"Size\"})\n",
    "        except : \n",
    "            print(\"Error in the format of the chr_size file\")\n",
    "            print(\"The chr_size file must have the following columns : Chrom, Size\")\n",
    "            return False\n",
    "    else :\n",
    "        size = chr_size_file\n",
    "\n",
    "    if type(annotation_file) == str : \n",
    "        dask_df = dd.read_csv(annotation_file,sep=\";\",  comment = \"#\",\n",
    "                        dtype={\"ensembl\":\"object\",\"id\":\"object\",\"name\":\"object\",\"ontology.group\":\"object\",\"gene.name\":\"object\",\"symbol\":\"object\"},\n",
    "                        usecols=[\"id\",\"name\",\"symbol\"],low_memory=False)\n",
    "        ann = dask_df.compute()\n",
    "        ann = ann[ann['id'].str.match('^GO.*')== True]\n",
    "    elif type(annotation_file) == pd.DataFrame : \n",
    "        ann = annotation_file.iloc[:,:4]\n",
    "        colname = list(ann.columns)\n",
    "        try : \n",
    "            ann = ann.rename(columns={colname[0]:\"id\",colname[1]:\"name\",colname[3]:\"symbol\"})\n",
    "        except : \n",
    "            print(\"Error in the format of the annotation file\")\n",
    "            print(\"The annotation file must have the following columns : id, name, symbol\")\n",
    "            return False\n",
    "    else :\n",
    "        ann = annotation_file\n",
    "\n",
    "    return test_data,regdom,size,ann\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "from numba import njit\n",
    "import Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext cython "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%cython\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "def select_id(ann,id) : \n",
    "    id_np = ann[\"id\"].to_numpy()\n",
    "    name_np = ann[\"name\"].to_numpy()\n",
    "    symbol_np = ann[\"symbol\"].to_numpy()\n",
    "    idx_np = np.where(id_np == id)\n",
    "    id_np = np.take(id_np,idx_np)[0]\n",
    "    name_np = np.take(name_np,idx_np,axis=0)[0]\n",
    "    symbol_np = np.take(symbol_np,idx_np,axis=0)[0]\n",
    "    return pd.DataFrame({\"id\":list(id_np),\"name\":list(name_np),\"symbol\":list(symbol_np)})\n",
    "\n",
    "def select_symbol(ann,symbol) : \n",
    "    id_np = ann[\"id\"].to_numpy()\n",
    "    name_np = ann[\"name\"].to_numpy()\n",
    "    symbol_np = ann[\"symbol\"].to_numpy()\n",
    "    idx_np = np.where(symbol_np == symbol)\n",
    "    id_np = np.take(id_np,idx_np)[0]\n",
    "    name_np = np.take(name_np,idx_np,axis=0)[0]\n",
    "    symbol_np = np.take(symbol_np,idx_np,axis=0)[0]\n",
    "    return pd.DataFrame({\"id\":list(id_np),\"name\":list(name_np),\"symbol\":list(symbol_np)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%cython\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def select_id(ann,id) : \n",
    "    id_np = ann[\"id\"].to_numpy()\n",
    "    name_np = ann[\"name\"].to_numpy()\n",
    "    symbol_np = ann[\"symbol\"].to_numpy()\n",
    "    idx_np = np.where(id_np == id)\n",
    "    id_np = np.take(id_np,idx_np)[0]\n",
    "    name_np = np.take(name_np,idx_np,axis=0)[0]\n",
    "    symbol_np = np.take(symbol_np,idx_np,axis=0)[0]\n",
    "    return pd.DataFrame({\"id\":list(id_np),\"name\":list(name_np),\"symbol\":list(symbol_np)})\n",
    "\n",
    "def select_symbol(ann,symbol) : \n",
    "    id_np = ann[\"id\"].to_numpy()\n",
    "    name_np = ann[\"name\"].to_numpy()\n",
    "    symbol_np = ann[\"symbol\"].to_numpy()\n",
    "    idx_np = np.where(symbol_np == symbol)\n",
    "    id_np = np.take(id_np,idx_np)[0]\n",
    "    name_np = np.take(name_np,idx_np,axis=0)[0]\n",
    "    symbol_np = np.take(symbol_np,idx_np,axis=0)[0]\n",
    "    return pd.DataFrame({\"id\":list(id_np),\"name\":list(name_np),\"symbol\":list(symbol_np)})\n",
    "\n",
    "def enrichment_binom_and_hypergeom(test,regdom,size,ann,asso) : \n",
    "    # Init Great\n",
    "    res = {}\n",
    "    hit={}\n",
    "\n",
    "    # init Hypergeom\n",
    "    hypergeom_gene_set=len(asso) # get the number of genes in the test gene set.\n",
    "    hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
    "\n",
    "    # Init binom \n",
    "    n_binom = test.shape[0]# get the number of genomic region in the test set\n",
    "    total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
    "\n",
    "    ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
    "    regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
    "    len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
    "\n",
    "    one = ann[:564427]\n",
    "    two = ann[564428:1128854]\n",
    "    three = ann[1128855:1693281]\n",
    "    four = ann[1693282:]\n",
    "\n",
    "    #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
    "    for name in asso :\n",
    "        # ann_name_gene = ann[ann[\"symbol\"].isin([name])]\n",
    "        tmp1 = select_symbol(one,name)\n",
    "        tmp2 = select_symbol(two,name); tmp2 = tmp2.reset_index(drop=True)\n",
    "        tmp3 = select_symbol(three,name); tmp3 = tmp3.reset_index(drop=True)\n",
    "        tmp4 = select_symbol(four,name); tmp4 = tmp4.reset_index(drop=True)\n",
    "        ann_name_gene = pd.concat([tmp1, tmp2,tmp3,tmp4],axis=0)\n",
    "\n",
    "        id = ann_name_gene[\"id\"]\n",
    "        tmp = []\n",
    "        for i in (list(id.unique())): \n",
    "            # gene_imply = ann[ann['id'].isin([i])]\n",
    "            tmp1 = select_id(one,i)\n",
    "            tmp2 = select_id(two,i); tmp2 = tmp2.reset_index(drop=True)\n",
    "            tmp3 = select_id(three,i); tmp3 = tmp3.reset_index(drop=True)\n",
    "            tmp4 = select_id(four,i); tmp4 = tmp4.reset_index(drop=True)\n",
    "            gene_imply = pd.concat([tmp1, tmp2,tmp3,tmp4],axis=0)\n",
    "            \n",
    "            K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
    "            curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
    "            k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation\n",
    "\n",
    "            if i not in list(hit.keys()) : \n",
    "                hit[i] = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "            k_binom=hit[i]\n",
    "            nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
    "            tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom))\n",
    "        res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu), hypergeom_cdf(hypergeom_total_number_gene,elem[4],hypergeom_gene_set,elem[5]) ] for elem in tmp})\n",
    "    return pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"binom_p_value\",2:\"hypergeom_p_value\"}).replace(0,np.nan).sort_values(by=\"binom_p_value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = \"../../data/tests/test_data/input/01_random.bed\"\n",
    "regdom = \"../../data/human/hg38/regulatory_domain.bed\"\n",
    "size = \"../../data/human/hg38/chr_size.bed\"\n",
    "t,r,s,a = loader(test,regdom,size,\"../../../ontologies_total.csv\")\n",
    "asso=get_association(t,r)\n",
    "test=t[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/icb/tom.mauger-birocheau/miniconda3/lib/python3.9/site-packages/line_profiler/ipython_extension.py:71: UserWarning: Could not extract a code object for the object <built-in function enrichment_binom_and_hypergeom>\n",
      "  profile = LineProfiler(*funcs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Timer unit: 1e-06 s"
     ]
    }
   ],
   "source": [
    "%lprun -f enrichment_binom_and_hypergeom enrichment_binom_and_hypergeom(test,r,s,a,asso)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Timer unit: 1e-06 s\n",
      "\n",
      "Total time: 213.928 s\n",
      "File: /tmp/ipykernel_11102/2426021515.py\n",
      "Function: enrichment_binom_and_hypergeom at line 1\n",
      "\n",
      "Line #      Hits         Time  Per Hit   % Time  Line Contents\n",
      "==============================================================\n",
      "     1                                           def enrichment_binom_and_hypergeom(test,regdom,size,ann,asso) : \n",
      "     2                                               # Init Great\n",
      "     3         1          4.0      4.0      0.0      res = {}\n",
      "     4         1          3.0      3.0      0.0      hit={}\n",
      "     5                                           \n",
      "     6                                               # init Hypergeom\n",
      "     7         1          4.0      4.0      0.0      hypergeom_gene_set=len(asso) # get the number of genes in the test gene set.\n",
      "     8         1         26.0     26.0      0.0      hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
      "     9                                           \n",
      "    10                                               # Init binom \n",
      "    11         1          8.0      8.0      0.0      n_binom = test.shape[0]# get the number of genomic region in the test set\n",
      "    12         1       1445.0   1445.0      0.0      total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
      "    13                                           \n",
      "    14         1      86046.0  86046.0      0.0      ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
      "    15         1     371924.0 371924.0      0.2      regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
      "    16         1      26359.0  26359.0      0.0      len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
      "    17                                           \n",
      "    18                                               \n",
      "    19                                           \n",
      "    20                                               #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
      "    21        59        103.0      1.7      0.0      for name in asso :\n",
      "    22                                                   # ann_name_gene = ann[ann[\"symbol\"].isin([name])]\n",
      "    23        58    2033595.0  35062.0      1.0          ann_name_gene = select_symbol(ann,name)\n",
      "    24        58       9856.0    169.9      0.0          id = ann_name_gene[\"id\"]\n",
      "    25        58        578.0     10.0      0.0          tmp = []\n",
      "    26      6201      15621.0      2.5      0.0          for i in (list(id.unique())): \n",
      "    27                                                       # gene_imply = ann[ann['id'].isin([i])]\n",
      "    28      6143  173076877.0  28174.7     80.9              gene_imply = select_id(ann,i)\n",
      "    29      6143      40004.0      6.5      0.0              K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
      "    30      6143   15532225.0   2528.4      7.3              curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
      "    31      6143    4572961.0    744.4      2.1              k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation\n",
      "    32                                           \n",
      "    33      6143     283217.0     46.1      0.1              if i not in list(hit.keys()) : \n",
      "    34      2060    2959830.0   1436.8      1.4                  hit[i] = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
      "    35      6143       8463.0      1.4      0.0              k_binom=hit[i]\n",
      "    36      6143    5163488.0    840.5      2.4              nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
      "    37      6143    1114758.0    181.5      0.5              tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom))\n",
      "    38        58    8517038.0 146845.5      4.0          res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu), hypergeom_cdf(hypergeom_total_number_gene,elem[4],hypergeom_gene_set,elem[5]) ] for elem in tmp})\n",
      "    39         1     113923.0 113923.0      0.1      return pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"binom_p_value\",2:\"hypergeom_p_value\"}).replace(0,np.nan).sort_values(by=\"binom_p_value\")"
     ]
    }
   ],
   "source": [
    "%lprun -f enrichment_binom_and_hypergeom enrichment_binom_and_hypergeom(test,r,s,a,asso)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "dd3aa23b046397ccf7f1ffb7cb82646c26b15f0b12555a0fca8fb427d9fb147f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
