{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example of combined use of bindome and greatpy "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook shows how to combine the bindome package and greatpy. For this it is necessary to have a file annotations which is in /storage/groups/ml01/datasets/. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The test will be done on the SRF file in order to obtain the same results as in the publication of great:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext line_profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import greatpy as great\n",
    "import bindome as bd\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>coordinate</th>\n",
       "      <th>summit.start</th>\n",
       "      <th>summit.end</th>\n",
       "      <th>k.summit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629831</td>\n",
       "      <td>629954</td>\n",
       "      <td>SRF:K-562</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>629938</td>\n",
       "      <td>629939</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:629831-629954</td>\n",
       "      <td>629838</td>\n",
       "      <td>630038</td>\n",
       "      <td>chr1:629838-630038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778597</td>\n",
       "      <td>778779</td>\n",
       "      <td>SRF:Ishikawa,GM12878,A-673-clone-Asp114,K-562,...</td>\n",
       "      <td>8</td>\n",
       "      <td>.</td>\n",
       "      <td>778671</td>\n",
       "      <td>778672</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:778597-778779</td>\n",
       "      <td>778571</td>\n",
       "      <td>778771</td>\n",
       "      <td>chr1:778571-778771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869817</td>\n",
       "      <td>870104</td>\n",
       "      <td>SRF:A-673-clone-Asp114</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>870004</td>\n",
       "      <td>870005</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:869817-870104</td>\n",
       "      <td>869904</td>\n",
       "      <td>870104</td>\n",
       "      <td>chr1:869904-870104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904714</td>\n",
       "      <td>904924</td>\n",
       "      <td>SRF:GM12878,A-673-clone-Asp114</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>904795</td>\n",
       "      <td>904796</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:904714-904924</td>\n",
       "      <td>904695</td>\n",
       "      <td>904895</td>\n",
       "      <td>chr1:904695-904895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>905143</td>\n",
       "      <td>905666</td>\n",
       "      <td>SRF:A-673-clone-Asp114</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>905393</td>\n",
       "      <td>905394</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:905143-905666</td>\n",
       "      <td>905293</td>\n",
       "      <td>905493</td>\n",
       "      <td>chr1:905293-905493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130753</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19567211</td>\n",
       "      <td>19567311</td>\n",
       "      <td>SRF:Hep-G2</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>19567231</td>\n",
       "      <td>19567232</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:19567211-19567311</td>\n",
       "      <td>19567131</td>\n",
       "      <td>19567331</td>\n",
       "      <td>chrY:19567131-19567331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130754</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19744932</td>\n",
       "      <td>19745043</td>\n",
       "      <td>SRF:Hep-G2</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>19744983</td>\n",
       "      <td>19744984</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:19744932-19745043</td>\n",
       "      <td>19744883</td>\n",
       "      <td>19745083</td>\n",
       "      <td>chrY:19744883-19745083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130755</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575583</td>\n",
       "      <td>20575822</td>\n",
       "      <td>SRF:WA01,Hep-G2</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>20575707</td>\n",
       "      <td>20575708</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:20575583-20575822</td>\n",
       "      <td>20575607</td>\n",
       "      <td>20575807</td>\n",
       "      <td>chrY:20575607-20575807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130756</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575934</td>\n",
       "      <td>20576114</td>\n",
       "      <td>SRF:WA01,Hep-G2</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>20576046</td>\n",
       "      <td>20576047</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:20575934-20576114</td>\n",
       "      <td>20575946</td>\n",
       "      <td>20576146</td>\n",
       "      <td>chrY:20575946-20576146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130757</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21021882</td>\n",
       "      <td>21022040</td>\n",
       "      <td>SRF:Hep-G2</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>21021946</td>\n",
       "      <td>21021947</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:21021882-21022040</td>\n",
       "      <td>21021846</td>\n",
       "      <td>21022046</td>\n",
       "      <td>chrY:21021846-21022046</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130758 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end  \\\n",
       "0       chr1    629831    629954   \n",
       "1       chr1    778597    778779   \n",
       "2       chr1    869817    870104   \n",
       "3       chr1    904714    904924   \n",
       "4       chr1    905143    905666   \n",
       "...      ...       ...       ...   \n",
       "130753  chrY  19567211  19567311   \n",
       "130754  chrY  19744932  19745043   \n",
       "130755  chrY  20575583  20575822   \n",
       "130756  chrY  20575934  20576114   \n",
       "130757  chrY  21021882  21022040   \n",
       "\n",
       "                                                        3  4  5         6  \\\n",
       "0                                               SRF:K-562  1  .    629938   \n",
       "1       SRF:Ishikawa,GM12878,A-673-clone-Asp114,K-562,...  8  .    778671   \n",
       "2                                  SRF:A-673-clone-Asp114  1  .    870004   \n",
       "3                          SRF:GM12878,A-673-clone-Asp114  2  .    904795   \n",
       "4                                  SRF:A-673-clone-Asp114  1  .    905393   \n",
       "...                                                   ... .. ..       ...   \n",
       "130753                                         SRF:Hep-G2  1  .  19567231   \n",
       "130754                                         SRF:Hep-G2  1  .  19744983   \n",
       "130755                                    SRF:WA01,Hep-G2  2  .  20575707   \n",
       "130756                                    SRF:WA01,Hep-G2  2  .  20576046   \n",
       "130757                                         SRF:Hep-G2  1  .  21021946   \n",
       "\n",
       "               7            8              coordinate  summit.start  \\\n",
       "0         629939  140,112,224      chr1:629831-629954        629838   \n",
       "1         778672  140,112,224      chr1:778597-778779        778571   \n",
       "2         870005  140,112,224      chr1:869817-870104        869904   \n",
       "3         904796  140,112,224      chr1:904714-904924        904695   \n",
       "4         905394  140,112,224      chr1:905143-905666        905293   \n",
       "...          ...          ...                     ...           ...   \n",
       "130753  19567232  140,112,224  chrY:19567211-19567311      19567131   \n",
       "130754  19744984  140,112,224  chrY:19744932-19745043      19744883   \n",
       "130755  20575708  140,112,224  chrY:20575583-20575822      20575607   \n",
       "130756  20576047  140,112,224  chrY:20575934-20576114      20575946   \n",
       "130757  21021947  140,112,224  chrY:21021882-21022040      21021846   \n",
       "\n",
       "        summit.end                k.summit  \n",
       "0           630038      chr1:629838-630038  \n",
       "1           778771      chr1:778571-778771  \n",
       "2           870104      chr1:869904-870104  \n",
       "3           904895      chr1:904695-904895  \n",
       "4           905493      chr1:905293-905493  \n",
       "...            ...                     ...  \n",
       "130753    19567331  chrY:19567131-19567331  \n",
       "130754    19745083  chrY:19744883-19745083  \n",
       "130755    20575807  chrY:20575607-20575807  \n",
       "130756    20576146  chrY:20575946-20576146  \n",
       "130757    21022046  chrY:21021846-21022046  \n",
       "\n",
       "[130758 rows x 13 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annpath = \"/home/tom/Stage/annotation\"\n",
    "# annpath='/storage/groups/ml01/datasets/annotations'\n",
    "bd.constants.ANNOTATIONS_DIRECTORY = annpath\n",
    "df1 = bd.bindome.datasets.REMAP2020.get_remap_peaks('SRF')\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Path should be modify when I have access to servers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1010 SRF:Ishikawa,GM12878,A-673-clone-Asp114,K-562,WA01,MCF-7,HCT-116,Hep-G2\n",
      "1416 SRF:GM12878,A-673-clone-Asp114\n",
      "629 SRF:Ishikawa,GM12878\n",
      "578 SRF:Ishikawa,MCF-7\n",
      "711 SRF:A-673-clone-Asp114,K-562\n",
      "959 SRF:A-673-clone-Asp114,Hep-G2\n",
      "1149 SRF:WA01\n",
      "646 SRF:A-673-clone-Asp114,HCT-116\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "name = pd.unique(list(df1[3]))\n",
    "for n in name : \n",
    "    if df1[df1[3]==n].shape[0] > 450 and df1[df1[3]==n].shape[0] < 2000 : \n",
    "        print(df1[df1[3]==n].shape[0],n)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the biosample used in the publication of great (556 genomic regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Chr_start</th>\n",
       "      <th>Chr_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1157504</td>\n",
       "      <td>1157660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2586482</td>\n",
       "      <td>2586614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3900438</td>\n",
       "      <td>3900894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6785102</td>\n",
       "      <td>6785311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10796590</td>\n",
       "      <td>10796818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130119</th>\n",
       "      <td>chrX</td>\n",
       "      <td>132023226</td>\n",
       "      <td>132023393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130187</th>\n",
       "      <td>chrX</td>\n",
       "      <td>133418786</td>\n",
       "      <td>133418958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130266</th>\n",
       "      <td>chrX</td>\n",
       "      <td>135344596</td>\n",
       "      <td>135344830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130562</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152830686</td>\n",
       "      <td>152830951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130645</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154000071</td>\n",
       "      <td>154000286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>578 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chr  Chr_start    Chr_end\n",
       "47      chr1    1157504    1157660\n",
       "178     chr1    2586482    2586614\n",
       "241     chr1    3900438    3900894\n",
       "371     chr1    6785102    6785311\n",
       "708     chr1   10796590   10796818\n",
       "...      ...        ...        ...\n",
       "130119  chrX  132023226  132023393\n",
       "130187  chrX  133418786  133418958\n",
       "130266  chrX  135344596  135344830\n",
       "130562  chrX  152830686  152830951\n",
       "130645  chrX  154000071  154000286\n",
       "\n",
       "[578 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df1[df1[3]==\"SRF:Ishikawa,MCF-7\"].iloc[:,0:3]\n",
    "df1 = df1.rename(columns={\"chr\":'Chr','start':\"Chr_start\",'end':\"Chr_end\"})\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_csv('~/Stage/bindome_df.bed',sep='\\t', index=False,header=False, columns=['Chr','Chr_start','Chr_end'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "test to try to understand the computing time problem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = great.tl.enrichment(\n",
    "    test = df1,\n",
    "    regdom_file=\"~/Stage/greatpy/data/human/regulatory_domain.bed\",\n",
    "    chr_size_file=\"~/Stage/greatpy/data/human/chr_size.bed\",\n",
    "    annotation=\"~/Stage/greatpy/data/human/ontologies.csv\",\n",
    "    binom=True,\n",
    "    hypergeom=True,\n",
    "    alpha=0.05,\n",
    "    correction=(0,0),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go_term</th>\n",
       "      <th>binom_p_value</th>\n",
       "      <th>hypergeom_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0070268</th>\n",
       "      <td>cornification</td>\n",
       "      <td>2.16672e-284</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  go_term  binom_p_value  hypergeom_p_value\n",
       "GO:0070268  cornification   2.16672e-284                NaN"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "9 min to analysis 500 genomic region with greatpy.tl.enrichment but the result is not the same as great"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Timer unit: 1e-06 s\n",
      "\n",
      "Total time: 763.606 s\n",
      "File: /home/tom/.local/lib/python3.8/site-packages/greatpy/tl/basic.py\n",
      "Function: enrichment at line 318\n",
      "\n",
      "Line #      Hits         Time  Per Hit   % Time  Line Contents\n",
      "==============================================================\n",
      "   318                                           def enrichment(test:str or pd.DataFrame,regdom_file,chr_size_file,annotation,binom=True,hypergeom=True,alpha=0.05,correction=(\"fdr\",0.05),sort_by=None): \n",
      "   319                                               # Data import \n",
      "   320         1         79.0     79.0      0.0      if not binom and not hypergeom : \n",
      "   321                                                   return False\n",
      "   322                                               \n",
      "   323         2      51765.0  25882.5      0.0      regdom = pd.read_csv(regdom_file,sep=\"\\t\",comment=\"#\",\n",
      "   324         1          4.0      4.0      0.0                      names=[\"Chr\", \"Chr_Start\", \"Chr_End\",\"Name\",\"tss\",\"Strand\"],dtype={\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\",\"Name\":\"object\",\"tss\":\"int64\",\"Strand\":\"object\"})\n",
      "   325                                           \n",
      "   326         1          4.0      4.0      0.0      if type(test) == str : \n",
      "   327                                                   test = pd.read_csv(test,sep=\"\\t\",comment=\"#\",\n",
      "   328                                                                   names=[\"Chr\", \"Chr_Start\", \"Chr_End\"],dtype={\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\"})\n",
      "   329                                               else : \n",
      "   330         1       1184.0   1184.0      0.0          test = test.iloc[:,:3]\n",
      "   331         1         31.0     31.0      0.0          colname = list(test.columns)\n",
      "   332         1          3.0      3.0      0.0          try : \n",
      "   333         1       2197.0   2197.0      0.0              test = test.rename(columns={colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\"})\n",
      "   334                                                   except : \n",
      "   335                                                       print(\"Error in test dataframe, please check your input\")\n",
      "   336                                                       print(\"Columns should be : chr...(type object), start(type int), end(type int)\")\n",
      "   337                                                       return False\n",
      "   338                                               \n",
      "   339         2       2489.0   1244.5      0.0      size = pd.read_csv(chr_size_file,sep=\"\\t\",comment=\"#\",\n",
      "   340         1          3.0      3.0      0.0                      names=[\"Chrom\",\"Size\"],dtype={\"Chrom\":\"object\", \"Size\":\"int64\"})\n",
      "   341                                           \n",
      "   342         2      26558.0  13279.0      0.0      dask_df = dd.read_csv(\"../data/human/ontologies.csv\",sep=\";\",  comment = \"#\",\n",
      "   343         1          4.0      4.0      0.0                      dtype={\"ensembl\":\"object\",\"id\":\"object\",\"name\":\"object\",\"ontology.group\":\"object\",\"gene.name\":\"object\",\"symbol\":\"object\"},\n",
      "   344         1          2.0      2.0      0.0                      usecols=[\"id\",\"name\",\"gene.name\",\"symbol\"],low_memory=False)\n",
      "   345         1     594167.0 594167.0      0.1      ann = dask_df.compute()\n",
      "   346                                               # ann = pd.read_csv(annotation,sep=\";\",  \n",
      "   347                                               #                 names=[\"ensembl\",\"id\",\"name\",\"ontology.group\",\"gene.name\",\"symbol\"],dtype={\"ensembl\":\"object\",\"id\":\"object\",\"name\":\"object\",\"ontology.group\":\"object\",\"gene.name\":\"object\",\"symbol\":\"object\"},\n",
      "   348                                               #                 usecols=[\"id\",\"name\",\"gene.name\",\"symbol\"],low_memory=False)\n",
      "   349         1     294338.0 294338.0      0.0      ann = ann[ann['id'].str.match('^GO.*')== True]\n",
      "   350                                           \n",
      "   351         1          3.0      3.0      0.0      if binom and hypergeom : \n",
      "   352                                                   # Init Great\n",
      "   353         1          2.0      2.0      0.0          res = {}\n",
      "   354         1          2.0      2.0      0.0          hit={}\n",
      "   355         1         66.0     66.0      0.0          hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
      "   356         1          6.0      6.0      0.0          n_binom = test.shape[0]# get the number of genomic region in the test set\n",
      "   357         1       1176.0   1176.0      0.0          total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
      "   358         1    2439624.0 2439624.0      0.3          asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
      "   359         1      17776.0  17776.0      0.0          ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
      "   360         1     122583.0 122583.0      0.0          regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
      "   361         1      26023.0  26023.0      0.0          len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
      "   362                                           \n",
      "   363                                                   # init Hypergeom\n",
      "   364         1          3.0      3.0      0.0          hypergeom_gene_set=len(asso) # get the number of genes in the test gene set.\n",
      "   365                                           \n",
      "   366                                                   #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
      "   367      1086       3108.0      2.9      0.0          for name in asso :\n",
      "   368      1085    7799300.0   7188.3      1.0              ann_name_gene = ann[ann[\"symbol\"].isin([name])]\n",
      "   369      1085      96686.0     89.1      0.0              id = ann_name_gene[\"id\"]\n",
      "   370      1085       5193.0      4.8      0.0              tmp = []\n",
      "   371     14844     154657.0     10.4      0.0              for i in (list(id.unique())): \n",
      "   372     13759  122270267.0   8886.6     16.0                  gene_imply = ann[ann['id'].isin([i])]\n",
      "   373     13759      97826.0      7.1      0.0                  K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
      "   374     13759   30651504.0   2227.7      4.0                  curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"gene.name\"]))]\n",
      "   375     13759   10639470.0    773.3      1.4                  k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation\n",
      "   376     13759     964815.0     70.1      0.1                  if i not in list(hit.keys()) : \n",
      "   377      4362  501049478.0 114866.9     65.6                      hit[i] = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
      "   378     13759      46864.0      3.4      0.0                  k_binom=hit[i]\n",
      "   379                                                               # k_binom = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
      "   380     13759    3603338.0    261.9      0.5                  nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
      "   381                                                           \n",
      "   382     13759    2662998.0    193.5      0.3                  tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom))\n",
      "   383                                                       # res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu), sum([hg.pmf(i,hypergeom_total_number_gene,hypergeom_gene_set,elem[4]) for i in range(elem[5],min(elem[4],hypergeom_gene_set)+1)]) ] for elem in tmp})\n",
      "   384     14844      47226.0      3.2      0.0              for elem in tmp : \n",
      "   385     13759      52895.0      3.8      0.0                  res[elem[2]] = [elem[3]]\n",
      "   386     13759     386549.0     28.1      0.1                  res[elem[2]].append(get_binom_pval(n_binom,elem[0],elem[1]/total_nu))\n",
      "   387                                                           # res[elem[2]].append(sum([hg.pmf(i,hypergeom_total_number_gene,hypergeom_gene_set,elem[4]) for i in range(elem[5],min(elem[4],hypergeom_gene_set)+1)]))\n",
      "   388     13759   79239279.0   5759.1     10.4                  res[elem[2]].append(hypergeom_cdf(hypergeom_total_number_gene,elem[4],hypergeom_gene_set,elem[5]))\n",
      "   389                                           \n",
      "   390         1     254139.0 254139.0      0.0          df= pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"binom_p_value\",2:\"hypergeom_p_value\"}).replace(0,np.nan)\n",
      "   391         1          4.0      4.0      0.0          if correction == (0,0) or correction[0] not in ['bonferroni','fdr'] or correction[1] >= 1 or correction[1] <= 0: \n",
      "   392         1          2.0      2.0      0.0              return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   393                                           \n",
      "   394                                                   elif correction[0] == \"bonferroni\" : \n",
      "   395                                                       df[\"binom_bonferroni_correction\"] = multipletests(df[\"binom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   396                                                       df[\"hypergeom_bonferroni_correction\"] = multipletests(df[\"hypergeom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   397                                                       df = df.loc[df[\"binom_bonferroni_correction\"] <= alpha]\n",
      "   398                                           \n",
      "   399                                                   elif correction[0] == \"fdr\" : \n",
      "   400                                                       df[\"binom_fdr_correction\"] = fdrcorrection(df[\"binom_p_value\"], alpha=correction[1])[1]\n",
      "   401                                                       df[\"hypergeom_fdr_correction\"] = fdrcorrection(df[\"hypergeom_p_value\"], alpha=correction[1])[1]\n",
      "   402                                                       df = df.loc[df[\"binom_fdr_correction\"] <= alpha]\n",
      "   403                                                   return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   404                                           \n",
      "   405                                               elif binom : \n",
      "   406                                                   # Init Great\n",
      "   407                                                   res = {}\n",
      "   408                                                   n_binom = test.shape[0]# get the number of genomic region in the test set\n",
      "   409                                                   total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
      "   410                                                   asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
      "   411                                                   ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
      "   412                                                   regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
      "   413                                                   len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
      "   414                                           \n",
      "   415                                                   #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
      "   416                                                   for name in asso :\n",
      "   417                                                       ann_name_gene = ann[ann[\"symbol\"] == name]\n",
      "   418                                                       id = ann_name_gene[\"id\"]\n",
      "   419                                                       tmp=[]\n",
      "   420                                                       for i in (list(id.unique())): \n",
      "   421                                                           gene_imply = ann[ann['id']==i]\n",
      "   422                                                           curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"gene.name\"]))]\n",
      "   423                                                           k_binom = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
      "   424                                                           nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
      "   425                                                           \n",
      "   426                                                           tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"]))\n",
      "   427                                                       res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu) ] for elem in tmp})\n",
      "   428                                                   df= pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"binom_p_value\"}).sort_values(by=\"binom_p_value\")\n",
      "   429                                                   if correction == (0,0) or correction[0] not in ['bonferroni','fdr'] or correction[1] >= 1 or correction[1]<=0: \n",
      "   430                                                       return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   431                                           \n",
      "   432                                                   elif correction[0] == \"bonferroni\" : \n",
      "   433                                                       df[\"binom_bonferroni_correction\"] = multipletests(df[\"binom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   434                                                       df = df.loc[df[\"binom_bonferroni_correction\"]<=alpha]\n",
      "   435                                           \n",
      "   436                                                   elif correction[0] == \"fdr\" :\n",
      "   437                                                       df[\"binom_fdr_correction\"] = fdrcorrection(df[\"binom_p_value\"], alpha=correction[1])[1]\n",
      "   438                                                       df = df.loc[df[\"binom_fdr_correction\"]<=alpha] \n",
      "   439                                                   return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   440                                           \n",
      "   441                                               else : \n",
      "   442                                                   # Init Great\n",
      "   443                                                   res = {}\n",
      "   444                                                   hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
      "   445                                                   asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
      "   446                                                   ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
      "   447                                                   regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
      "   448                                                   len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
      "   449                                           \n",
      "   450                                                   # init Hypergeom\n",
      "   451                                                   hypergeom_gene_set = len(asso) # get the number of genes in the test gene set.\n",
      "   452                                           \n",
      "   453                                                   #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
      "   454                                                   for name in asso :\n",
      "   455                                                       ann_name_gene = ann[ann[\"symbol\"] == name]\n",
      "   456                                                       id = ann_name_gene[\"id\"]\n",
      "   457                                                       tmp = []\n",
      "   458                                                       for i in (list(id.unique())): \n",
      "   459                                                           gene_imply = ann[ann['id']==i]\n",
      "   460                                                           K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
      "   461                                                           curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"gene.name\"]))]\n",
      "   462                                                           k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation                \n",
      "   463                                                           tmp.append((i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom)) \n",
      "   464                                                       res.update({elem[0]:[ elem[1], sum([hg.pmf(i,hypergeom_total_number_gene,hypergeom_gene_set,elem[2]) for i in range(elem[3],min(elem[2],hypergeom_gene_set)+1)]) ] for elem in tmp}) \n",
      "   465                                           \n",
      "   466                                                   df = pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"hypergeom_p_value\"}).sort_values(by=\"hypergeom_p_value\")\n",
      "   467                                                   \n",
      "   468                                                   if correction == (0,0) or correction[0] not in ['bonferroni','fdr'] or correction[1] >= 1 or correction[1]<=0: \n",
      "   469                                                       df = df.loc[df[\"hypergeom_p_value\"] <= alpha]\n",
      "   470                                                       return df.sort_values(by=sort_by) if sort_by != None else df  \n",
      "   471                                           \n",
      "   472                                                   elif correction[0] == \"bonferroni\" : \n",
      "   473                                                       df[\"hypergeom_bonferroni_correction\"] = multipletests(df[\"hypergeom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   474                                                       df = df.loc[df[\"hypergeom_bonferroni_correction\"] <= alpha]\n",
      "   475                                           \n",
      "   476                                                   elif correction[0] == \"fdr\" : \n",
      "   477                                                       df[\"hypergeom_fdr_correction\"] = fdrcorrection(df[\"hypergeom_p_value\"], alpha=correction[1])[1]\n",
      "   478                                                       df = df.loc[df[\"hypergeom_fdr_correction\"] <= alpha]\n",
      "   479                                           \n",
      "   480                                                   return df.sort_values(by=sort_by) if sort_by != None else df"
     ]
    }
   ],
   "source": [
    "%lprun -f great.tl.enrichment great.tl.enrichment(df1,\"~/Stage/greatpy/data/human/regulatory_domain.bed\",\"~/Stage/greatpy/data/human/chr_size.bed\",\"~/Stage/greatpy/data/human/ontologies.csv\",True,True,0.05,(0,0))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
