{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example of combined use of bindome and greatpy "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook shows how to combine the bindome package and greatpy. For this it is necessary to have a file annotations which is in /storage/groups/ml01/datasets/. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The test will be done on the SRF file in order to obtain the same results as in the publication of great:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "The line_profiler extension is already loaded. To reload it, use:\n",
      "  %reload_ext line_profiler\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext line_profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import greatpy as great\n",
    "import bindome as bd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>coordinate</th>\n",
       "      <th>summit.start</th>\n",
       "      <th>summit.end</th>\n",
       "      <th>k.summit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629831</td>\n",
       "      <td>629954</td>\n",
       "      <td>SRF:K-562</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>629938</td>\n",
       "      <td>629939</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:629831-629954</td>\n",
       "      <td>629838</td>\n",
       "      <td>630038</td>\n",
       "      <td>chr1:629838-630038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778597</td>\n",
       "      <td>778779</td>\n",
       "      <td>SRF:Ishikawa,GM12878,A-673-clone-Asp114,K-562,...</td>\n",
       "      <td>8</td>\n",
       "      <td>.</td>\n",
       "      <td>778671</td>\n",
       "      <td>778672</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:778597-778779</td>\n",
       "      <td>778571</td>\n",
       "      <td>778771</td>\n",
       "      <td>chr1:778571-778771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869817</td>\n",
       "      <td>870104</td>\n",
       "      <td>SRF:A-673-clone-Asp114</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>870004</td>\n",
       "      <td>870005</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:869817-870104</td>\n",
       "      <td>869904</td>\n",
       "      <td>870104</td>\n",
       "      <td>chr1:869904-870104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904714</td>\n",
       "      <td>904924</td>\n",
       "      <td>SRF:GM12878,A-673-clone-Asp114</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>904795</td>\n",
       "      <td>904796</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:904714-904924</td>\n",
       "      <td>904695</td>\n",
       "      <td>904895</td>\n",
       "      <td>chr1:904695-904895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>905143</td>\n",
       "      <td>905666</td>\n",
       "      <td>SRF:A-673-clone-Asp114</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>905393</td>\n",
       "      <td>905394</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chr1:905143-905666</td>\n",
       "      <td>905293</td>\n",
       "      <td>905493</td>\n",
       "      <td>chr1:905293-905493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130753</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19567211</td>\n",
       "      <td>19567311</td>\n",
       "      <td>SRF:Hep-G2</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>19567231</td>\n",
       "      <td>19567232</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:19567211-19567311</td>\n",
       "      <td>19567131</td>\n",
       "      <td>19567331</td>\n",
       "      <td>chrY:19567131-19567331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130754</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19744932</td>\n",
       "      <td>19745043</td>\n",
       "      <td>SRF:Hep-G2</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>19744983</td>\n",
       "      <td>19744984</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:19744932-19745043</td>\n",
       "      <td>19744883</td>\n",
       "      <td>19745083</td>\n",
       "      <td>chrY:19744883-19745083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130755</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575583</td>\n",
       "      <td>20575822</td>\n",
       "      <td>SRF:WA01,Hep-G2</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>20575707</td>\n",
       "      <td>20575708</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:20575583-20575822</td>\n",
       "      <td>20575607</td>\n",
       "      <td>20575807</td>\n",
       "      <td>chrY:20575607-20575807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130756</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575934</td>\n",
       "      <td>20576114</td>\n",
       "      <td>SRF:WA01,Hep-G2</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>20576046</td>\n",
       "      <td>20576047</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:20575934-20576114</td>\n",
       "      <td>20575946</td>\n",
       "      <td>20576146</td>\n",
       "      <td>chrY:20575946-20576146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130757</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21021882</td>\n",
       "      <td>21022040</td>\n",
       "      <td>SRF:Hep-G2</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>21021946</td>\n",
       "      <td>21021947</td>\n",
       "      <td>140,112,224</td>\n",
       "      <td>chrY:21021882-21022040</td>\n",
       "      <td>21021846</td>\n",
       "      <td>21022046</td>\n",
       "      <td>chrY:21021846-21022046</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130758 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end  \\\n",
       "0       chr1    629831    629954   \n",
       "1       chr1    778597    778779   \n",
       "2       chr1    869817    870104   \n",
       "3       chr1    904714    904924   \n",
       "4       chr1    905143    905666   \n",
       "...      ...       ...       ...   \n",
       "130753  chrY  19567211  19567311   \n",
       "130754  chrY  19744932  19745043   \n",
       "130755  chrY  20575583  20575822   \n",
       "130756  chrY  20575934  20576114   \n",
       "130757  chrY  21021882  21022040   \n",
       "\n",
       "                                                        3  4  5         6  \\\n",
       "0                                               SRF:K-562  1  .    629938   \n",
       "1       SRF:Ishikawa,GM12878,A-673-clone-Asp114,K-562,...  8  .    778671   \n",
       "2                                  SRF:A-673-clone-Asp114  1  .    870004   \n",
       "3                          SRF:GM12878,A-673-clone-Asp114  2  .    904795   \n",
       "4                                  SRF:A-673-clone-Asp114  1  .    905393   \n",
       "...                                                   ... .. ..       ...   \n",
       "130753                                         SRF:Hep-G2  1  .  19567231   \n",
       "130754                                         SRF:Hep-G2  1  .  19744983   \n",
       "130755                                    SRF:WA01,Hep-G2  2  .  20575707   \n",
       "130756                                    SRF:WA01,Hep-G2  2  .  20576046   \n",
       "130757                                         SRF:Hep-G2  1  .  21021946   \n",
       "\n",
       "               7            8              coordinate  summit.start  \\\n",
       "0         629939  140,112,224      chr1:629831-629954        629838   \n",
       "1         778672  140,112,224      chr1:778597-778779        778571   \n",
       "2         870005  140,112,224      chr1:869817-870104        869904   \n",
       "3         904796  140,112,224      chr1:904714-904924        904695   \n",
       "4         905394  140,112,224      chr1:905143-905666        905293   \n",
       "...          ...          ...                     ...           ...   \n",
       "130753  19567232  140,112,224  chrY:19567211-19567311      19567131   \n",
       "130754  19744984  140,112,224  chrY:19744932-19745043      19744883   \n",
       "130755  20575708  140,112,224  chrY:20575583-20575822      20575607   \n",
       "130756  20576047  140,112,224  chrY:20575934-20576114      20575946   \n",
       "130757  21021947  140,112,224  chrY:21021882-21022040      21021846   \n",
       "\n",
       "        summit.end                k.summit  \n",
       "0           630038      chr1:629838-630038  \n",
       "1           778771      chr1:778571-778771  \n",
       "2           870104      chr1:869904-870104  \n",
       "3           904895      chr1:904695-904895  \n",
       "4           905493      chr1:905293-905493  \n",
       "...            ...                     ...  \n",
       "130753    19567331  chrY:19567131-19567331  \n",
       "130754    19745083  chrY:19744883-19745083  \n",
       "130755    20575807  chrY:20575607-20575807  \n",
       "130756    20576146  chrY:20575946-20576146  \n",
       "130757    21022046  chrY:21021846-21022046  \n",
       "\n",
       "[130758 rows x 13 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annpath = \"/home/tom/Stage/annotation\"\n",
    "# annpath='/storage/groups/ml01/datasets/annotations'\n",
    "bd.constants.ANNOTATIONS_DIRECTORY = annpath\n",
    "df1 = bd.bindome.datasets.REMAP2020.get_remap_peaks('SRF')\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1010 SRF:Ishikawa,GM12878,A-673-clone-Asp114,K-562,WA01,MCF-7,HCT-116,Hep-G2\n",
      "1416 SRF:GM12878,A-673-clone-Asp114\n",
      "629 SRF:Ishikawa,GM12878\n",
      "578 SRF:Ishikawa,MCF-7\n",
      "711 SRF:A-673-clone-Asp114,K-562\n",
      "959 SRF:A-673-clone-Asp114,Hep-G2\n",
      "1149 SRF:WA01\n",
      "646 SRF:A-673-clone-Asp114,HCT-116\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "name = pd.unique(list(df1[3]))\n",
    "for n in name : \n",
    "    if df1[df1[3]==n].shape[0] > 450 and df1[df1[3]==n].shape[0] < 2000 : \n",
    "        print(df1[df1[3]==n].shape[0],n)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the biosample used in the publication of great (556 genomic regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = df1[df1[3]==\"SRF:Ishikawa,MCF-7\"]\n",
    "df1 = df1.rename(columns={\"chr\":'Chr','start':\"Chr_start\",'end':\"Chr_end\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Path should be modify when I have access to servers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_csv('~/Stage/bindome_df.bed',sep='\\t', index=False,header=False, columns=['Chr','Chr_start','Chr_end'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "test to try to understand the computing time problem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_10_row = df1[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Timer unit: 1e-06 s\n",
      "\n",
      "Total time: 13.1051 s\n",
      "File: /home/tom/.local/lib/python3.8/site-packages/greatpy/tl/basic.py\n",
      "Function: enrichment at line 277\n",
      "\n",
      "Line #      Hits         Time  Per Hit   % Time  Line Contents\n",
      "==============================================================\n",
      "   277                                           def enrichment(test:str or pd.DataFrame,regdom_file,chr_size_file,annotation,binom=True,hypergeom=True,alpha=0.05,correction=(\"fdr\",0.05),sort_by=None): \n",
      "   278                                               # Data import \n",
      "   279         1         10.0     10.0      0.0      if not binom and not hypergeom : \n",
      "   280                                                   return False\n",
      "   281                                               \n",
      "   282         2      32649.0  16324.5      0.2      regdom = pd.read_csv(regdom_file,sep=\"\\t\",comment=\"#\",\n",
      "   283         1          6.0      6.0      0.0                      names=[\"Chr\", \"Chr_Start\", \"Chr_End\",\"Name\",\"tss\",\"Strand\"],dtype={\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\",\"Name\":\"object\",\"tss\":\"int64\",\"Strand\":\"object\"})\n",
      "   284                                           \n",
      "   285         1          5.0      5.0      0.0      if type(test) == str : \n",
      "   286                                                   test = pd.read_csv(test,sep=\"\\t\",comment=\"#\",\n",
      "   287                                                                   names=[\"Chr\", \"Chr_Start\", \"Chr_End\"],dtype={\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\"})\n",
      "   288                                               else : \n",
      "   289         1        458.0    458.0      0.0          test = test.iloc[:,:3]\n",
      "   290         1         13.0     13.0      0.0          colname = list(test.columns)\n",
      "   291         1          2.0      2.0      0.0          try : \n",
      "   292         1        796.0    796.0      0.0              test = test.rename(columns={colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\"})\n",
      "   293                                                   except : \n",
      "   294                                                       print(\"Error in test dataframe, please check your input\")\n",
      "   295                                                       print(\"Columns should be : chr...(type object), start(type int), end(type int)\")\n",
      "   296                                                       return False\n",
      "   297                                               \n",
      "   298         2       4404.0   2202.0      0.0      size = pd.read_csv(chr_size_file,sep=\"\\t\",comment=\"#\",\n",
      "   299         1          3.0      3.0      0.0                      names=[\"Chrom\",\"Size\"],dtype={\"Chrom\":\"object\", \"Size\":\"int64\"})\n",
      "   300                                           \n",
      "   301         2     402105.0 201052.5      3.1      ann = pd.read_csv(annotation,sep=\";\",  \n",
      "   302         1          3.0      3.0      0.0                      names=[\"ensembl\",\"id\",\"name\",\"ontology.group\",\"gene.name\",\"symbol\"],dtype={\"ensembl\":\"object\",\"id\":\"object\",\"name\":\"object\",\"ontology.group\":\"object\",\"gene.name\":\"object\",\"symbol\":\"object\"},\n",
      "   303         1          2.0      2.0      0.0                      usecols=[\"id\",\"name\",\"gene.name\",\"symbol\"],low_memory=True)\n",
      "   304         1     262692.0 262692.0      2.0      ann = ann[ann['id'].str.match('^GO.*')== True]\n",
      "   305                                           \n",
      "   306         1          3.0      3.0      0.0      if binom and hypergeom : \n",
      "   307                                                   # Init Great\n",
      "   308         1          3.0      3.0      0.0          res = {}\n",
      "   309         1         10.0     10.0      0.0          hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
      "   310         1          5.0      5.0      0.0          n_binom = test.shape[0]# get the number of genomic region in the test set\n",
      "   311         1        401.0    401.0      0.0          total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
      "   312         1      51219.0  51219.0      0.4          asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
      "   313         1      18251.0  18251.0      0.1          ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
      "   314         1      55267.0  55267.0      0.4          regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
      "   315         1      27329.0  27329.0      0.2          len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
      "   316                                           \n",
      "   317                                                   # init Hypergeom\n",
      "   318         1          5.0      5.0      0.0          hypergeom_gene_set=len(asso) # get the number of genes in the test gene set.\n",
      "   319                                           \n",
      "   320                                                   #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
      "   321        21         75.0      3.6      0.0          for name in asso :\n",
      "   322        20     466841.0  23342.0      3.6              ann_name_gene = ann[ann[\"symbol\"]==name]\n",
      "   323        20       2190.0    109.5      0.0              id = ann_name_gene[\"id\"]\n",
      "   324        20         81.0      4.0      0.0              tmp = []\n",
      "   325       170       3335.0     19.6      0.0              for i in (list(id.unique())): \n",
      "   326       150    4052756.0  27018.4     30.9                  gene_imply = ann[ann['id']==i]\n",
      "   327       150       1535.0     10.2      0.0                  K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
      "   328       150     419087.0   2793.9      3.2                  curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"gene.name\"]))]\n",
      "   329       150     127212.0    848.1      1.0                  k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation\n",
      "   330       150    3088908.0  20592.7     23.6                  k_binom = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
      "   331       150      68188.0    454.6      0.5                  nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
      "   332                                                           \n",
      "   333       150      31774.0    211.8      0.2                  tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom))\n",
      "   334        20    3977258.0 198862.9     30.3              res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu), sum([hg.pmf(i,hypergeom_total_number_gene,hypergeom_gene_set,elem[4]) for i in range(elem[5],min(elem[4],hypergeom_gene_set)+1)]) ] for elem in tmp})\n",
      "   335                                                   \n",
      "   336         1      10191.0  10191.0      0.1          df= pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"binom_p_value\",2:\"hypergeom_p_value\"})\n",
      "   337         1          3.0      3.0      0.0          if correction == (0,0) or correction[0] not in ['bonferroni','fdr'] or correction[1] >= 1 or correction[1] <= 0: \n",
      "   338         1          3.0      3.0      0.0              return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   339                                           \n",
      "   340                                                   elif correction[0] == \"bonferroni\" : \n",
      "   341                                                       df[\"binom_bonferroni_correction\"] = multipletests(df[\"binom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   342                                                       df[\"hypergeom_bonferroni_correction\"] = multipletests(df[\"hypergeom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   343                                                       df = df.loc[df[\"binom_bonferroni_correction\"] <= alpha]\n",
      "   344                                           \n",
      "   345                                                   elif correction[0] == \"fdr\" : \n",
      "   346                                                       df[\"binom_fdr_correction\"] = fdrcorrection(df[\"binom_p_value\"], alpha=correction[1])[1]\n",
      "   347                                                       df[\"hypergeom_fdr_correction\"] = fdrcorrection(df[\"hypergeom_p_value\"], alpha=correction[1])[1]\n",
      "   348                                                       df = df.loc[df[\"binom_fdr_correction\"] <= alpha]\n",
      "   349                                                   return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   350                                           \n",
      "   351                                               elif binom : \n",
      "   352                                                   # Init Great\n",
      "   353                                                   res = {}\n",
      "   354                                                   n_binom = test.shape[0]# get the number of genomic region in the test set\n",
      "   355                                                   total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
      "   356                                                   asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
      "   357                                                   ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
      "   358                                                   regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
      "   359                                                   len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
      "   360                                           \n",
      "   361                                                   #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
      "   362                                                   for name in asso :\n",
      "   363                                                       ann_name_gene = ann[ann[\"symbol\"] == name]\n",
      "   364                                                       id = ann_name_gene[\"id\"]\n",
      "   365                                                       tmp=[]\n",
      "   366                                                       for i in (list(id.unique())): \n",
      "   367                                                           gene_imply = ann[ann['id']==i]\n",
      "   368                                                           curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"gene.name\"]))]\n",
      "   369                                                           k_binom = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
      "   370                                                           nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
      "   371                                                           \n",
      "   372                                                           tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"]))\n",
      "   373                                                       res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu) ] for elem in tmp})\n",
      "   374                                                   df= pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"binom_p_value\"}).sort_values(by=\"binom_p_value\")\n",
      "   375                                                   if correction == (0,0) or correction[0] not in ['bonferroni','fdr'] or correction[1] >= 1 or correction[1]<=0: \n",
      "   376                                                       return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   377                                           \n",
      "   378                                                   elif correction[0] == \"bonferroni\" : \n",
      "   379                                                       df[\"binom_bonferroni_correction\"] = multipletests(df[\"binom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   380                                                       df = df.loc[df[\"binom_bonferroni_correction\"]<=alpha]\n",
      "   381                                           \n",
      "   382                                                   elif correction[0] == \"fdr\" :\n",
      "   383                                                       df[\"binom_fdr_correction\"] = fdrcorrection(df[\"binom_p_value\"], alpha=correction[1])[1]\n",
      "   384                                                       df = df.loc[df[\"binom_fdr_correction\"]<=alpha] \n",
      "   385                                                   return df.sort_values(by=sort_by) if sort_by != None else df \n",
      "   386                                           \n",
      "   387                                               else : \n",
      "   388                                                   # Init Great\n",
      "   389                                                   res = {}\n",
      "   390                                                   hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
      "   391                                                   asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
      "   392                                                   ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
      "   393                                                   regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
      "   394                                                   len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
      "   395                                           \n",
      "   396                                                   # init Hypergeom\n",
      "   397                                                   hypergeom_gene_set = len(asso) # get the number of genes in the test gene set.\n",
      "   398                                           \n",
      "   399                                                   #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
      "   400                                                   for name in asso :\n",
      "   401                                                       ann_name_gene = ann[ann[\"symbol\"] == name]\n",
      "   402                                                       id = ann_name_gene[\"id\"]\n",
      "   403                                                       tmp = []\n",
      "   404                                                       for i in (list(id.unique())): \n",
      "   405                                                           gene_imply = ann[ann['id']==i]\n",
      "   406                                                           K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
      "   407                                                           curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"gene.name\"]))]\n",
      "   408                                                           k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation                \n",
      "   409                                                           tmp.append((i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom)) \n",
      "   410                                                       res.update({elem[0]:[ elem[1], sum([hg.pmf(i,hypergeom_total_number_gene,hypergeom_gene_set,elem[2]) for i in range(elem[3],min(elem[2],hypergeom_gene_set)+1)]) ] for elem in tmp}) \n",
      "   411                                           \n",
      "   412                                                   df = pd.DataFrame(res).transpose().rename(columns={0:\"go_term\",1:\"hypergeom_p_value\"}).sort_values(by=\"hypergeom_p_value\")\n",
      "   413                                                   \n",
      "   414                                                   if correction == (0,0) or correction[0] not in ['bonferroni','fdr'] or correction[1] >= 1 or correction[1]<=0: \n",
      "   415                                                       df = df.loc[df[\"hypergeom_p_value\"] <= alpha]\n",
      "   416                                                       return df.sort_values(by=sort_by) if sort_by != None else df  \n",
      "   417                                           \n",
      "   418                                                   elif correction[0] == \"bonferroni\" : \n",
      "   419                                                       df[\"hypergeom_bonferroni_correction\"] = multipletests(df[\"hypergeom_p_value\"], alpha=correction[1], method='bonferroni')[1]\n",
      "   420                                                       df = df.loc[df[\"hypergeom_bonferroni_correction\"] <= alpha]\n",
      "   421                                           \n",
      "   422                                                   elif correction[0] == \"fdr\" : \n",
      "   423                                                       df[\"hypergeom_fdr_correction\"] = fdrcorrection(df[\"hypergeom_p_value\"], alpha=correction[1])[1]\n",
      "   424                                                       df = df.loc[df[\"hypergeom_fdr_correction\"] <= alpha]\n",
      "   425                                           \n",
      "   426                                                   return df.sort_values(by=sort_by) if sort_by != None else df"
     ]
    }
   ],
   "source": [
    "%lprun -f great.tl.enrichment great.tl.enrichment(df_10_row,\"~/Stage/greatpy/data/human/regulatory_domain.bed\",\"~/Stage/greatpy/data/human/chr_size.bed\",\"~/Stage/greatpy/data/human/ontologies.csv\",True,True,0.05,(0,0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = great.tl.enrichment(\n",
    "#     test = '~/Stage/bindome_df.bed',\n",
    "#     regdom_file=\"~/Stage/greatpy/data/human/regulatory_domain.bed\",\n",
    "#     chr_size_file=\"~/Stage/greatpy/data/human/chr_size.bed\",\n",
    "#     annotation=\"~/Stage/greatpy/data/human/ontologies.csv\",\n",
    "#     binom=True,\n",
    "#     hypergeom=True,\n",
    "#     alpha=0.05,\n",
    "#     correction=(\"fdr\",0.05),\n",
    "#     sort_by=\"binom_fdr_correction\",\n",
    "#     )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Computation time to much longer, more than 3 hours locally"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
