{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext line_profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import greatpy as great"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = \"../data/tests/test_data/input/01_random.bed\"\n",
    "regdom = \"../data/human/hg38/regulatory_domain.bed\"\n",
    "size = \"../data/human/hg38/chr_size.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "from math import lgamma, log, exp,fabs\n",
    "from scipy.special import comb\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import dask.dataframe as dd \n",
    "pd.options.display.float_format = '{:12.5e}'.format\n",
    "\n",
    "class GREAT: \n",
    "    def loader(test_data:None or str or pd.DataFrame,regdom_file:None or str or pd.DataFrame,chr_size_file:None or str or pd.DataFrame,annotation_file:None or str or pd.DataFrame) : \n",
    "        \"\"\"\n",
    "        This function is used to load all datasets needed for the enrichment calculation\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test_data : None or str or pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom_file : None or str or pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size_file : None or str or pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation_file : None or str or pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        test_data : pd.DataFrame\n",
    "            Genomic set of peaks to be tested in the good format \n",
    "        regdom : pd.DataFrame\n",
    "            Regulatory domain of all genes in the genome in the good format\n",
    "        size : pd.DataFrame\n",
    "            Table with the size of each chromosome in the good format\n",
    "        ann : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome in the good format\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\n",
    "        ...    \"../../data/tests/test_data/input/02_srf_hg38.bed\",\n",
    "        ...    \"../../data/human/hg38/regulatory_domain.bed\",\n",
    "        ...    \"../../data/human/hg38/chr_size.bed\",\n",
    "        ...    \"../../data/human/ontologies.csv\"\n",
    "        ...    )\n",
    "\n",
    "        >>> test.head()\n",
    "            |    | Chr   |   Chr_Start |   Chr_End |\n",
    "            |---:|:------|------------:|----------:|\n",
    "            |  0 | chr1  |     1052028 |   1052049 |\n",
    "            |  1 | chr1  |     1065512 |   1065533 |\n",
    "            |  2 | chr1  |     1067375 |   1067397 |\n",
    "            |  3 | chr1  |     1068083 |   1068119 |\n",
    "            |  4 | chr1  |    10520283 |  10520490 |\n",
    "\n",
    "        >>> regdom.head()\n",
    "            |    | Chr   |   Chr_Start |   Chr_End | Name      |   tss | Strand   |\n",
    "            |---:|:------|------------:|----------:|:----------|------:|:---------|\n",
    "            |  0 | chr1  |           0 |     22436 | MIR6859-1 | 17436 | -        |\n",
    "            |  1 | chr1  |       16436 |     22436 | MIR6859-2 | 17436 | -        |\n",
    "            |  2 | chr1  |       16436 |     22436 | MIR6859-3 | 17436 | -        |\n",
    "            |  3 | chr1  |       16436 |     28370 | MIR6859-4 | 17436 | -        |\n",
    "            |  4 | chr1  |       22436 |     34370 | WASH7P    | 29370 | -        |\n",
    "\n",
    "        >>> size.head()\n",
    "            |    | Chrom   |      Size |\n",
    "            |---:|:--------|----------:|\n",
    "            |  0 | chr1    | 248956422 |\n",
    "            |  1 | chr2    | 242193529 |\n",
    "            |  2 | chr3    | 198295559 |\n",
    "            |  3 | chr4    | 190214555 |\n",
    "            |  4 | chr5    | 181538259 |\n",
    "            \n",
    "        >>> ann.head()\n",
    "            |    | id         | name                                                   | symbol        |\n",
    "            |---:|:-----------|:-------------------------------------------------------|:--------------|\n",
    "            |  0 | GO:0003924 | GTPase activity                                        | DNAJC25-GNG10 |\n",
    "            |  1 | GO:0007186 | G protein-coupled receptor signaling pathway           | DNAJC25-GNG10 |\n",
    "            |  2 | GO:0003723 | RNA binding                                            | NUDT4B        |\n",
    "            |  3 | GO:0005829 | cytosol                                                | NUDT4B        |\n",
    "            |  4 | GO:0008486 | diphosphoinositol-polyphosphate diphosphatase activity | NUDT4B        |\n",
    "        \n",
    "        \"\"\"\n",
    "\n",
    "        if type(regdom_file) == str :\n",
    "            regdom = pd.read_csv(regdom_file,sep = \"\\t\",comment=\"#\",\n",
    "                        names = [\"Chr\", \"Chr_Start\", \"Chr_End\",\"Name\",\"tss\",\"Strand\"],\n",
    "                        dtype = {\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\",\"Name\":\"object\",\"tss\":\"int64\",\"Strand\":\"object\"})\n",
    "        elif type(regdom_file) == pd.DataFrame :\n",
    "            regdom = regdom_file.iloc[:,:6]\n",
    "            colname = list(regdom.columns)\n",
    "            try : \n",
    "                regdom = regdom.rename(columns = {colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\",colname[3]:\"Name\",colname[4]:\"tss\",colname[5]:\"Strand\"})\n",
    "            except :\n",
    "                print(\"Error in the format of the regdom file\")\n",
    "                print(\"The regdom file must have the following columns : Chr, Chr_Start, Chr_End, Name, tss, Strand\")\n",
    "                return False \n",
    "        else : \n",
    "            regdom = regdom_file\n",
    "\n",
    "        if type(test_data) == str : \n",
    "            test_data = pd.read_csv(test_data,sep = \"\\t\",comment = \"#\",usecols = [0,1,2],\n",
    "                            names = [\"Chr\", \"Chr_Start\", \"Chr_End\"],\n",
    "                            dtype = {\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\"})\n",
    "        elif type(test_data) == pd.DataFrame : \n",
    "            test_data = test_data.iloc[:,:3]\n",
    "            colname = list(test_data.columns)\n",
    "            try : \n",
    "                test_data = test_data.rename(columns={colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\"})\n",
    "            except : \n",
    "                print(\"Error in test dataframe, please check your input\")\n",
    "                print(\"Columns should be : chr...(type object), start(type int), end(type int)\")\n",
    "                return False\n",
    "        else :\n",
    "            pass\n",
    "\n",
    "        if type(chr_size_file) == str :\n",
    "            size = pd.read_csv(chr_size_file,sep = \"\\t\",comment = \"#\",\n",
    "                            names = [\"Chrom\",\"Size\"],\n",
    "                            dtype = {\"Chrom\":\"object\", \"Size\":\"int64\"})\n",
    "        elif type(chr_size_file) == pd.DataFrame :\n",
    "            size = chr_size_file.iloc[:,:2]\n",
    "            colname = list(size.columns)\n",
    "            try : \n",
    "                size = size.rename(columns = {colname[0]:\"Chrom\",colname[1]:\"Size\"})\n",
    "            except : \n",
    "                print(\"Error in the format of the chr_size file\")\n",
    "                print(\"The chr_size file must have the following columns : Chrom, Size\")\n",
    "                return False\n",
    "        else :\n",
    "            size = chr_size_file\n",
    "\n",
    "        if type(annotation_file) == str : \n",
    "            dask_df = dd.read_csv(annotation_file,sep = \";\",  comment = \"#\",\n",
    "                            dtype = {\"ensembl\":\"object\",\"id\":\"object\",\"name\":\"object\",\"ontology.group\":\"object\",\"gene.name\":\"object\",\"symbol\":\"object\"},\n",
    "                            usecols = [\"id\",\"name\",\"symbol\"],low_memory = False)\n",
    "            ann = dask_df.compute()\n",
    "            ann = ann[ann['id'].str.match('^GO.*')== True]\n",
    "        elif type(annotation_file) == pd.DataFrame : \n",
    "            ann = annotation_file.iloc[:,:4]\n",
    "            colname = list(ann.columns)\n",
    "            try : \n",
    "                ann = ann.rename(columns={colname[0]:\"id\",colname[1]:\"name\",colname[3]:\"symbol\"})\n",
    "            except : \n",
    "                print(\"Error in the format of the annotation file\")\n",
    "                print(\"The annotation file must have the following columns : id, name, symbol\")\n",
    "                return False\n",
    "        else :\n",
    "            ann = annotation_file\n",
    "\n",
    "        return test_data,regdom,size,ann\n",
    "\n",
    "    def __enrichment_binom_and_hypergeom(test:pd.DataFrame, regdom:pd.DataFrame, size: pd.DataFrame, ann:pd.DataFrame, asso:list) -> pd.DataFrame : \n",
    "        \"\"\"\n",
    "        This private function is used to compute the enrichment of the test data using the binomial test and the hypergeometric test.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom : pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size :  pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        asso : list \n",
    "            List of the association between gene from regdom and peaks from test\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the binomial test and the the hypergeometric test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/tests/test_data/input/02_srf_hg38.bed\",\"../../data/human/hg38/regulatory_domain.bed\",\"../../data/human/hg38/chr_size.bed\",\"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.____enrichment_binom_and_hypergeom(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    size = size,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom)\n",
    "        ...    )\n",
    "\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  |\n",
    "        \"\"\"\n",
    "        # Init Great\n",
    "        res = {}\n",
    "        hit = {}\n",
    "\n",
    "        # init Hypergeom\n",
    "        hypergeom_gene_set = len(asso) # get the number of genes in the test gene set.\n",
    "        hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
    "\n",
    "        # Init binom \n",
    "        n_binom = test.shape[0]# get the number of genomic region in the test set\n",
    "        total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
    "\n",
    "        ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
    "        regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
    "        len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
    "\n",
    "        #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
    "        for name in asso :\n",
    "            ann_name_gene = ann[ann[\"symbol\"].isin([name])]\n",
    "            id = ann_name_gene[\"id\"]\n",
    "            tmp = []\n",
    "            for i in (list(id.unique())) : \n",
    "                gene_imply = ann[ann['id'].isin([i])]\n",
    "                K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
    "                curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
    "                k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation\n",
    "\n",
    "                if i not in list(hit.keys()) : \n",
    "                    hit[i] = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "                k_binom = hit[i]\n",
    "                nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
    "                tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom))\n",
    "\n",
    "            res.update({elem[2] : [ \n",
    "                elem[3],\n",
    "                get_binom_pval(n_binom,elem[0],elem[1]/total_nu),\n",
    "                elem[0]/(elem[1]/total_nu), #binom enrichment \n",
    "                hypergeom_cdf(hypergeom_total_number_gene,elem[4],hypergeom_gene_set,elem[5]),\n",
    "                (elem[5]*hypergeom_total_number_gene)/(hypergeom_gene_set*elem[4]) # Hypergeom enrichment \n",
    "                ] for elem in tmp})\n",
    "\n",
    "        return pd.DataFrame(res).transpose().rename(columns = {0 : \"go_term\",1 : \"binom_p_value\",2 : \"binom_fold_enrichment\",3 : \"hypergeom_p_value\",4 : \"hypergeometric_fold_enrichment\"}).replace(0,np.nan).dropna().sort_values(by = \"binom_p_value\")\n",
    "    \n",
    "    def __enrichment_binom(test:pd.DataFrame, regdom:pd.DataFrame, size: pd.DataFrame, ann:pd.DataFrame, asso:list) -> pd.DataFrame :\n",
    "        \"\"\"\n",
    "        This private function is used to compute the enrichment of the test data using the binomial test.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom : pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size :  pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        asso : list \n",
    "            List of the association between gene from regdom and peaks from test\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the binomial test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/tests/test_data/input/02_srf_hg38.bed\",\"../../data/human/hg38/regulatory_domain.bed\",\"../../data/human/hg38/chr_size.bed\",\"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.____enrichment_binom(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    size = size,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom)\n",
    "        ...    )\n",
    "\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |\n",
    "        \"\"\" \n",
    "        # Init Great\n",
    "        res = {}\n",
    "        hit = {}\n",
    "\n",
    "        # Init binom \n",
    "        n_binom = test.shape[0]# get the number of genomic region in the test set\n",
    "        total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
    "        \n",
    "        ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
    "        regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
    "        len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
    "\n",
    "        #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
    "        for name in asso :\n",
    "            ann_name_gene = ann[ann[\"symbol\"].isin([name])]\n",
    "            id = ann_name_gene[\"id\"]\n",
    "            tmp = []\n",
    "            for i in (list(id.unique())) : \n",
    "                gene_imply = ann[ann['id'].isin([i])]\n",
    "                curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
    "\n",
    "                if i not in list(hit.keys()) : \n",
    "                    hit[i] = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "                k_binom = hit[i]\n",
    "                nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
    "                tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"]))\n",
    "\n",
    "            res.update({elem[2] : [ \n",
    "                elem[3],\n",
    "                get_binom_pval(n_binom,elem[0],elem[1]/total_nu),\n",
    "                elem[0]/(elem[1]/total_nu) # binom enrichment \n",
    "                ] for elem in tmp})\n",
    "\n",
    "        return pd.DataFrame(res).transpose().rename(columns = {0 : \"go_term\",1 : \"binom_p_value\",2 : \"binom_fold_enrichment\"}).sort_values(by = \"binom_p_value\")\n",
    "\n",
    "    def __enrichment_hypergeom(test:pd.DataFrame,regdom:pd.DataFrame,ann:pd.DataFrame,asso:list) -> pd.DataFrame : \n",
    "        \"\"\"\n",
    "        This private function is used to compute the enrichment of the test data using the hypergeometric test.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom : pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size :  pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        asso : list \n",
    "            List of the association between gene from regdom and peaks from test\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the hypergeometric test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.____enrichment_hypergeom(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom)\n",
    "        ...    )\n",
    "\n",
    "        >>> enrichment.head()   \n",
    "        \"\"\" \n",
    "        # Init Great\n",
    "        res = {}\n",
    "\n",
    "        # Init hypergeom\n",
    "        hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome\n",
    "        hypergeom_gene_set = len(asso) # get the number of genes in the test gene set.\n",
    "\n",
    "        ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
    "        regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
    "\n",
    "        #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
    "        for name in asso :\n",
    "            ann_name_gene = ann[ann[\"symbol\"] == name]\n",
    "            id = ann_name_gene[\"id\"]\n",
    "            tmp = []\n",
    "            for i in (list(id.unique())) : \n",
    "                gene_imply = ann[ann['id']==i]\n",
    "                K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
    "                curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
    "                k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation                \n",
    "                tmp.append((i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom))\n",
    "\n",
    "            res.update({elem[0]:[\n",
    "                elem[1], \n",
    "                hypergeom_cdf(hypergeom_total_number_gene,elem[2],hypergeom_gene_set,elem[3]),\n",
    "                (elem[3]*hypergeom_total_number_gene)/(hypergeom_gene_set*elem[2]) # hypergeom enrichment \n",
    "                ] for elem in tmp}) \n",
    "\n",
    "        return pd.DataFrame(res).transpose().rename(columns = {0 : \"go_term\",1 : \"hypergeom_p_value\",2 : \"hypergeometric_fold_enrichment\"}).replace(0,np.nan).dropna().sort_values(by = \"hypergeom_p_value\")\n",
    "\n",
    "\n",
    "    def enrichment(test_file: str or pd.DataFrame,regdom_file: str or pd.DataFrame,chr_size_file: str or pd.DataFrame, annotation_file: str or pd.DataFrame, binom=True,hypergeom=True) -> pd.DataFrame :\n",
    "        \"\"\"\n",
    "        This function is a wrapper of the 3 private methods: \n",
    "        * GREAT.__enrichment_binom_and_hypergeom \n",
    "        * GREAT.__enrichment_binom \n",
    "        * GREAT.__enrichment_hypergeom\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test_file : str or pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom_file : str or pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size_file : str or pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation_file : str or pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        binom : bool (default True)\n",
    "            If True, the binomial test is used.\n",
    "        hypergeom : bool (default True)\n",
    "            If True, the hypergeometric test is used.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the hypergeometric test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.enrichment(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    chr_size_file = size,\n",
    "        ...    ann = ann,\n",
    "        ...    binom=True,\n",
    "        ...    hypergeom=True\n",
    "        ...    )\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  |\n",
    "\n",
    "        >>> enrichment = GREAT.enrichment(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom),\n",
    "        ...    binom=True,\n",
    "        ...    hypergeom=False\n",
    "        ...    )\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |\n",
    "\n",
    "        >>> enrichment = GREAT.enrichment(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom),\n",
    "        ...    binom=False,\n",
    "        ...    hypergeom=True\n",
    "        ...    )\n",
    "        >>> enrichment.head()\n",
    "\n",
    "        \"\"\"\n",
    "        if not binom and not hypergeom : \n",
    "            return False\n",
    "        \n",
    "        test,regdom,size,ann = GREAT.loader(test_file,regdom_file,chr_size_file, annotation_file)\n",
    "        asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
    "\n",
    "\n",
    "        if binom and hypergeom : \n",
    "            return GREAT.__enrichment_binom_and_hypergeom(test,regdom,size,ann,asso)\n",
    "\n",
    "        elif binom : \n",
    "            return GREAT.__enrichment_binom(test,regdom,size,ann,asso)\n",
    "\n",
    "        else : \n",
    "              return GREAT.__enrichment_hypergeom(test,regdom,ann,asso)\n",
    "\n",
    "    def set_bonferroni(self,alpha:float=0.05) -> pd.DataFrame : \n",
    "        \"\"\"\n",
    "        This function create new columns in the dataframe with the Bonferroni correction\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        alpha : float\n",
    "            alpha value for the Bonferroni correction\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe new columns with the Bonferroni correction for each p-value\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = great.tl.GREAT.enrichment(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\",binom=True,hypergeom=True)\n",
    "        >>> bonferroni = GREAT.set_bonferroni(enrichment,alpha=0.05)\n",
    "        >>> bonferroni.head()\n",
    "        ...    |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |   binom_bonferroni |   hypergeom_bonferroni |\n",
    "        ...    |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|-------------------:|-----------------------:|\n",
    "        ...    | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  |        3.0754e-10  |                      1 |\n",
    "        ...    | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  |        2.87384e-07 |                      1 |\n",
    "        ...    | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  |        2.87384e-07 |                      1 |\n",
    "        ...    | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 |        3.74892e-07 |                      1 |\n",
    "        ...    | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  |        7.68931e-07 |                      1 |\n",
    "\n",
    "        \"\"\"\n",
    "        for col in self.columns : \n",
    "            if col in [\"binom_p_value\",\"hypergeom_p_value\"] : \n",
    "                col_split = col.split(\"_\")\n",
    "                self[f\"{col_split[0]}_bonferroni\"] = multipletests(self[col], alpha=alpha, method='bonferroni')[1]\n",
    "        return self \n",
    "\n",
    "    def set_fdr(self,alpha:float=0.05) -> pd.DataFrame : \n",
    "        \"\"\"\n",
    "        This function create new columns in the dataframe with the fdr correction\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        alpha : float\n",
    "            alpha value for the fdr correction\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe new columns with the fdr correction for each p-value\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = great.tl.GREAT.enrichment(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\",binom=True,hypergeom=True)\n",
    "        >>> fdr = GREAT.set_fdr(enrichment,alpha=0.05)\n",
    "        >>> fdr.head()\n",
    "        ...    |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |   binom_fdr |   hypergeom_fdr |\n",
    "        ...    |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|------------:|----------------:|\n",
    "        ...    | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  | 3.0754e-10  |       0.0913909 |\n",
    "        ...    | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  | 9.3723e-08  |       0.0913909 |\n",
    "        ...    | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  | 9.3723e-08  |       0.0913909 |\n",
    "        ...    | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 | 9.3723e-08  |       0.0913909 |\n",
    "        ...    | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  | 1.53786e-07 |       0.0913909 |\n",
    "\n",
    "        \"\"\"\n",
    "        for col in self.columns : \n",
    "            if col in [\"binom_p_value\",\"hypergeom_p_value\"] : \n",
    "                col_split = col.split(\"_\")\n",
    "                self[f\"{col_split[0]}_fdr\"] = multipletests(self[col], alpha=alpha, method='fdr_bh')[1]\n",
    "        return self \n",
    "\n",
    "    def set_threshold(self,colname:str, alpha:int=0.05) -> pd.DataFrame : \n",
    "        \"\"\"\n",
    "        This function allows to delete rows according to the p-value of the column taken as argument. By default the alpha value is 0.05\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        alpha : float\n",
    "            alpha value for the fdr correction\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe with the rows deleted according to the p-value threshold\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = great.tl.GREAT.enrichment(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\",binom=True,hypergeom=True)\n",
    "        >>> enrichment.shape[0]\n",
    "        ...    594\n",
    "\n",
    "        >>> significant = GREAT.set_threshold(enrichment,colname=\"binom_p_value\",alpha=0.05)\n",
    "        >>> significant.shape[0]\n",
    "        ...    310\n",
    "\n",
    "        \"\"\"\n",
    "        if colname in self.columns: \n",
    "            self = self.loc[self[colname] <= alpha]\n",
    "        return self \n",
    "\n",
    "\n",
    "######################################################################\n",
    "################# Utils function used by GREAT class #################\n",
    "######################################################################\n",
    "def get_association(test,regdom) -> list : \n",
    "    \"\"\"\n",
    "    Function allowing from a file of genomic regions from CHIPseq \n",
    "    and a file of genomic regulatory domains to determine the names \n",
    "    of genes associated with at least one genomic region \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    test : pd.dataFrame\n",
    "        df of the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    res : list\n",
    "        list of gene associated with at least with one test peak\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    test = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1052028],\n",
    "    ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']})\n",
    "\n",
    "    >>> get_association(test,regdom)        \n",
    "        ['RNF223']\n",
    "    \n",
    "    \"\"\"\n",
    "    res = []\n",
    "    for i in range(test.shape[0]) :\n",
    "        currTest = test.iloc[i]\n",
    "        regdom_curr_test = regdom.loc[(regdom[\"Chr\"] == currTest[\"Chr\"])].sort_values(\"Chr_Start\")\n",
    "        regdom_curr_test = regdom_curr_test.loc[\n",
    "            ((regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_End\"])) | # regdom overlap totally test \n",
    "            ((regdom_curr_test[\"Chr_Start\"] >= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] <= currTest[\"Chr_End\"])) | # test overlap totally regdom \n",
    "            ((regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] <= currTest[\"Chr_End\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_Start\"])) | # regdom overlap not totally test on left side \n",
    "            ((regdom_curr_test[\"Chr_Start\"] >= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_End\"]) & (regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_End\"])) # regdom overlap not totally test on right side \n",
    "            ] \n",
    "        res = res + list(regdom_curr_test[\"Name\"])\n",
    "    return list(dict.fromkeys(res))\n",
    "\n",
    "def len_regdom(regdom:pd.DataFrame) -> dict :  \n",
    "    \"\"\"\n",
    "    Function to calculate for each gene name from regdom the\n",
    "     size of the regulatory region for this gene in the genome \n",
    "\n",
    "    Parameters\n",
    "    ----------    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dict\n",
    "        dictionary in which each key corresponds to a gene name \n",
    "        from regdom and the value is the size of the regulatory \n",
    "        region for that gene\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']}))\n",
    "\n",
    "    >>> len_regdom(regdom)\n",
    "        {'RNF223': 80097, 'C1orf159': 52710}\n",
    "\n",
    "    \"\"\"\n",
    "    test = regdom[\"Chr_End\"] - regdom[\"Chr_Start\"]\n",
    "    return pd.DataFrame({\"len\" : list(test)},index = regdom[\"Name\"]).to_dict()[\"len\"]\n",
    "\n",
    "def number_of_hit(test,regdom) -> int : \n",
    "    \"\"\" \n",
    "    Function to calculate the number of hits from several \n",
    "    genomic regions and the file describing the regulatory regions\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    test : pd.dataFrame\n",
    "        df of the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    nb : int\n",
    "        number of hit \n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    test = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1052028],\n",
    "    ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']})\n",
    "\n",
    "    >>> number_of_hit(test,regdom)        \n",
    "        1\n",
    "    \n",
    "    \"\"\"\n",
    "    nb = 0\n",
    "    regdom = regdom[[\"Chr\",\"Chr_Start\",\"Chr_End\"]]\n",
    "    for i in range(test.shape[0]) : \n",
    "        chrom = test.iat[i,0]\n",
    "        start = test.iat[i,1]\n",
    "        end = test.iat[i,2]\n",
    "        regdom_np = regdom[\"Chr\"].to_numpy()\n",
    "        reg_start = regdom[\"Chr_Start\"].to_numpy()\n",
    "        reg_end = regdom[\"Chr_End\"].to_numpy()\n",
    "        Chr_reduce = np.where(regdom_np == chrom)\n",
    "        reg_start = np.take(reg_start,Chr_reduce,axis = 0)[0]\n",
    "        reg_end = np.take(reg_end,Chr_reduce,axis = 0)[0]\n",
    "\n",
    "        if any((reg_start <= start) & (reg_end >= end)):  \n",
    "            nb += 1\n",
    "    return nb\n",
    "\n",
    "def betacf(a,b,x) : \n",
    "    \"\"\" Used by betai: Evaluates continued fraction for incomplete beta function \"\"\"\n",
    "    maxit = 10000\n",
    "    eps = 3.0e-7 \n",
    "    fpmin = 1.0e-30\n",
    "    qab = a + b\n",
    "    qap = a + 1\n",
    "    qam = a - 1\n",
    "    c = 1\n",
    "    d = 1 - qab * x / qap\n",
    "    if fabs(d) < fpmin :\n",
    "        d = fpmin\n",
    "    d = 1 / d \n",
    "    h = d\n",
    "    for m in range(1,maxit + 1) : \n",
    "        m2 = 2 * m\n",
    "        aa = m * (b - m) * x / ((qam + m2) * (a + m2))\n",
    "        d = 1.0 + aa * d\n",
    "        if (fabs(d) < fpmin) : \n",
    "            d = fpmin\n",
    "        c = 1.0 + aa / c\n",
    "        if (fabs(c) < fpmin):\n",
    "            c = fpmin\n",
    "        d = 1.0 / d\n",
    "        h *= d * c\n",
    "        aa = -(a + m) * (qab + m) * x / ((a + m2) * (qap + m2))\n",
    "        d = 1.0 + aa * d  \n",
    "        if (fabs(d) < fpmin):\n",
    "            d = fpmin\n",
    "        c = 1.0 + aa / c\n",
    "        if (fabs(c) < fpmin):\n",
    "            c = fpmin\n",
    "        d = 1.0 / d\n",
    "        dell = d * c\n",
    "        h *= dell\n",
    "        if (fabs(dell-1.0) < eps):\n",
    "            break\n",
    "    if (m > maxit):\n",
    "        print(\"a or b too big, or MAXIT too small in betacf\")\n",
    "        return False\n",
    "    return h\n",
    "\n",
    "def betai(a,b,x) :\n",
    "    \"\"\"Returns the incomplete beta function Ix(a, b).\"\"\"\n",
    "    if x < 0 or x > 1 : \n",
    "        # print(\"bad x in routine betai\")\n",
    "        return False\n",
    "    if x == 0 or x == 1 : \n",
    "        bt = 0.0\n",
    "    else : \n",
    "        bt = exp(lgamma(a + b) - lgamma(a) - lgamma(b) + a * log(x) + b * log(1.0 - x))\n",
    "    if x < (a + 1) / (a + b + 2) : \n",
    "        return bt * betacf(a,b,x) / a\n",
    "    return 1 - bt * betacf(b,a,1-x) / b\n",
    "\n",
    "def get_binom_pval(n:int,k:int,p:float) -> float :\n",
    "    \"\"\"\n",
    "    This function allows to calculate the binomial probability \n",
    "    of obtaining k in a set of size n and whose probability is p \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n : int\n",
    "        Number of genomic region in the test set \n",
    "    k : int \n",
    "        Number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "    p : float\n",
    "        Percentage of genome annotated\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        binomial probability\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> get_binom_pval(100,2,0.2)\n",
    "        0.9999999947037065\n",
    "    \n",
    "    \"\"\"\n",
    "    if k == 0 : return 1\n",
    "    else : return betai(k , n - k + 1,p)\n",
    "\n",
    "def hypergeom_pmf(N:int, K:int, n:int, k:int) -> float :  \n",
    "    \"\"\"\n",
    "    Function to calculate the probability mass function for hypergeometric distribution\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    N : int\n",
    "        Total number of gene in the genome\n",
    "    K : int \n",
    "        Number of genes in the genome with annotation\n",
    "    n : int\n",
    "        Number of gene in the test set\n",
    "    k : int\n",
    "        Number of genes in the test gene set with annotation\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        proability mass function\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> hypergeom_pmf(100,10,30,1)\n",
    "        0.11270773995748315\n",
    "    \n",
    "    \"\"\"\n",
    "    Achoosex = comb(K,k,exact=True) \n",
    "    NAchoosenx = comb(N - K, n - k,exact=True) \n",
    "    Nchoosen = comb(N,n,exact=True) \n",
    "    return ((Achoosex) * NAchoosenx) / Nchoosen \n",
    "\n",
    "def hypergeom_cdf(N:int, K:int, n:int, k:int) -> float :\n",
    "    \"\"\"\n",
    "    Function to calculate the cumulative density funtion for hypergeometric distribution\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    N : int\n",
    "        Total number of gene in the genome\n",
    "    K : int \n",
    "        Number of genes in the genome with annotation\n",
    "    n : int\n",
    "        Number of gene in the test set\n",
    "    k : int\n",
    "        Number of genes in the test gene set with annotation\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        Cumulative density function\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> hypergeom_cdf(100,10,30,1)\n",
    "        0.9770827595419788\n",
    "    \n",
    "    \"\"\"\n",
    "    return np.sum([hypergeom_pmf(N, K, n, x) for x in range(k , min(K,n) + 1)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'shape'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb Cell 5\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=0'>1</a>\u001b[0m asso \u001b[39m=\u001b[39m get_association(test,regdom)\n",
      "\u001b[1;32m/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb Cell 5\u001b[0m in \u001b[0;36mget_association\u001b[0;34m(test, regdom)\u001b[0m\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=573'>574</a>\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=574'>575</a>\u001b[0m \u001b[39mFunction allowing from a file of genomic regions from CHIPseq \u001b[39;00m\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=575'>576</a>\u001b[0m \u001b[39mand a file of genomic regulatory domains to determine the names \u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=608'>609</a>\u001b[0m \u001b[39m\u001b[39;00m\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=609'>610</a>\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=610'>611</a>\u001b[0m res \u001b[39m=\u001b[39m []\n\u001b[0;32m--> <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=611'>612</a>\u001b[0m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(test\u001b[39m.\u001b[39;49mshape[\u001b[39m0\u001b[39m]) :\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=612'>613</a>\u001b[0m     currTest \u001b[39m=\u001b[39m test\u001b[39m.\u001b[39miloc[i]\n\u001b[1;32m    <a href='vscode-notebook-cell://ssh-remote%2Bicb-lisa/home/icb/tom.mauger-birocheau/internship/greatpy-1/notebooks/test.ipynb#ch0000009vscode-remote?line=613'>614</a>\u001b[0m     regdom_curr_test \u001b[39m=\u001b[39m regdom\u001b[39m.\u001b[39mloc[(regdom[\u001b[39m\"\u001b[39m\u001b[39mChr\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m currTest[\u001b[39m\"\u001b[39m\u001b[39mChr\u001b[39m\u001b[39m\"\u001b[39m])]\u001b[39m.\u001b[39msort_values(\u001b[39m\"\u001b[39m\u001b[39mChr_Start\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'shape'"
     ]
    }
   ],
   "source": [
    "asso = get_association(test,regdom) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Could not find module great.tl.GREAT.__enrichment_binom_and_hypergeom.\n",
      "AttributeError: type object 'GREAT' has no attribute '__enrichment_binom_and_hypergeom'\n"
     ]
    }
   ],
   "source": [
    "%lprun -f great.tl.GREAT.__enrichment_binom_and_hypergeom great.tl.GREAT.__enrichment_binom_and_hypergeom(test,regdom,size,\"../data/human/ontologies.csv\",asso)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "dd3aa23b046397ccf7f1ffb7cb82646c26b15f0b12555a0fca8fb427d9fb147f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
