{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computation of annotation enrichment in a genetic set "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code excerpt shows how to use the annotation enrichment computation function in a genetic set.\n",
    "The funciton return a dataframe of binom and hypergeometric p-value (<= 0.05) and for each p-value a fdr correction with alpha = 0.05.   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the second part of the code we will compare the results obtained with the Great webserver and between the binomial and hypergeometric probability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import greatpy as grt\n",
    "import pandas as pd\n",
    "from math import inf\n",
    "from numpy import log,nan, int64,cov,corrcoef\n",
    "from scipy.stats import pearsonr\n",
    "from seaborn import scatterplot as sp\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import rpy2\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects import r as r\n",
    "pandas2ri.activate()\n",
    "import pandas as pd\n",
    "\n",
    "import os \n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import time "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from math import lgamma, log, exp,fabs,inf\n",
    "from scipy.special import comb\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import dask.dataframe as dd \n",
    "import numpy as np \n",
    "\n",
    "pd.options.display.float_format = '{:12.5e}'.format\n",
    "\n",
    "def get_association(test,regdom): \n",
    "    \"\"\"\n",
    "    Function allowing from a file of genomic regions from CHIPseq \n",
    "    and a file of genomic regulatory domains to determine the names \n",
    "    of genes associated with at least one genomic region \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    test : pd.dataFrame\n",
    "        df of the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    res : list\n",
    "        list of gene associated with at least with one test peak\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    test = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1052028],\n",
    "    ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']})\n",
    "\n",
    "    >>> get_association(test,regdom)        \n",
    "        ['RNF223']\n",
    "    \n",
    "    \"\"\"\n",
    "    res = []\n",
    "    for i in range(test.shape[0]) :\n",
    "        currTest = test.iloc[i]\n",
    "        regdom_curr_test = regdom.loc[(regdom[\"Chr\"] == currTest[\"Chr\"])].sort_values(\"Chr_Start\")\n",
    "        regdom_curr_test = regdom_curr_test.loc[\n",
    "            ((regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_End\"])) | # regdom overlap totally test \n",
    "            ((regdom_curr_test[\"Chr_Start\"] >= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] <= currTest[\"Chr_End\"])) | # test overlap totally regdom \n",
    "            ((regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] <= currTest[\"Chr_End\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_Start\"])) | # regdom overlap not totally test on left side \n",
    "            ((regdom_curr_test[\"Chr_Start\"] >= currTest[\"Chr_Start\"]) & (regdom_curr_test[\"Chr_End\"] >= currTest[\"Chr_End\"]) & (regdom_curr_test[\"Chr_Start\"] <= currTest[\"Chr_End\"])) # regdom overlap not totally test on right side \n",
    "            ] \n",
    "        res = res + list(regdom_curr_test[\"Name\"])\n",
    "    return list(dict.fromkeys(res))\n",
    "\n",
    "def len_regdom(regdom:pd.DataFrame): \n",
    "    \"\"\"\n",
    "    Function to calculate for each gene name from regdom the\n",
    "     size of the regulatory region for this gene in the genome \n",
    "\n",
    "    Parameters\n",
    "    ----------    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dict\n",
    "        dictionary in which each key corresponds to a gene name \n",
    "        from regdom and the value is the size of the regulatory \n",
    "        region for that gene\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']}))\n",
    "\n",
    "    >>> len_regdom(regdom)\n",
    "        {'RNF223': 80097, 'C1orf159': 52710}\n",
    "\n",
    "    \"\"\"\n",
    "    test = regdom[\"Chr_End\"]-regdom[\"Chr_Start\"]\n",
    "    return pd.DataFrame({\"len\":list(test)},index = regdom[\"Name\"]).to_dict()[\"len\"]\n",
    "\n",
    "def number_of_hit(test,regdom): \n",
    "    \"\"\" \n",
    "    Function to calculate the number of hits from several \n",
    "    genomic regions and the file describing the regulatory regions\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    test : pd.dataFrame\n",
    "        df of the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "    \n",
    "    regdom : pd.dataFrame\n",
    "        df of the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    nb : int\n",
    "        number of hit \n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    test = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1052028],\n",
    "    ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "    regdom = pd.DataFrame({\n",
    "    ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "    ...    \"Chr_Start\":[1034992,1079306],\n",
    "    ...    \"Chr_End\": [1115089,1132016],\n",
    "    ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "    ...    \"tss\":[1074306,1116089],\n",
    "    ...    \"strand\":['-','-']})\n",
    "\n",
    "    >>> number_of_hit(test,regdom)        \n",
    "        1\n",
    "    \n",
    "    \"\"\"\n",
    "    nb = 0\n",
    "    regdom = regdom[[\"Chr\",\"Chr_Start\",\"Chr_End\"]]\n",
    "    for i in range(test.shape[0]) : \n",
    "        chrom = test.iat[i,0]\n",
    "        start = test.iat[i,1]\n",
    "        end = test.iat[i,2]\n",
    "        regdom_np = regdom[\"Chr\"].to_numpy()\n",
    "        reg_start = regdom[\"Chr_Start\"].to_numpy()\n",
    "        reg_end = regdom[\"Chr_End\"].to_numpy()\n",
    "        Chr_reduce = np.where(regdom_np == chrom)\n",
    "        reg_start = np.take(reg_start,Chr_reduce,axis = 0)[0]\n",
    "        reg_end = np.take(reg_end,Chr_reduce,axis = 0)[0]\n",
    "\n",
    "        if any((reg_start <= start) & (reg_end >= end)):  \n",
    "            nb += 1\n",
    "    return nb\n",
    "\n",
    "def betacf(a,b,x): \n",
    "    \"\"\" Used by betai: Evaluates continued fraction for incomplete beta function \"\"\"\n",
    "    maxit = 10000\n",
    "    eps = 3.0e-7 \n",
    "    fpmin = 1.0e-30\n",
    "    qab = a+b\n",
    "    qap = a+1\n",
    "    qam = a-1\n",
    "    c = 1\n",
    "    d = 1-qab*x/qap\n",
    "    if fabs(d) < fpmin :\n",
    "        d = fpmin\n",
    "    d = 1/d \n",
    "    h = d\n",
    "    for m in range(1,maxit+1) : \n",
    "        m2 = 2*m\n",
    "        aa = m*(b-m)*x / ((qam+m2) * (a+m2))\n",
    "        d = 1.0+aa*d\n",
    "        if (fabs(d) < fpmin) : \n",
    "            d = fpmin\n",
    "        c = 1.0+aa/c\n",
    "        if (fabs(c) < fpmin):\n",
    "            c = fpmin\n",
    "        d = 1.0/d\n",
    "        h *= d*c\n",
    "        aa = -(a+m) * (qab+m)*x / ((a+m2) * (qap+m2))\n",
    "        d = 1.0+aa*d  \n",
    "        if (fabs(d) < fpmin):\n",
    "            d = fpmin\n",
    "        c = 1.0+aa/c\n",
    "        if (fabs(c) < fpmin):\n",
    "            c = fpmin\n",
    "        d = 1.0/d\n",
    "        dell = d*c\n",
    "        h *= dell\n",
    "        if (fabs(dell-1.0) < eps):\n",
    "            break\n",
    "    if (m > maxit):\n",
    "        print(\"a or b too big, or MAXIT too small in betacf\")\n",
    "        return False\n",
    "    return h\n",
    "\n",
    "def betai(a,b,x):\n",
    "    \"\"\"Returns the incomplete beta function Ix(a, b).\"\"\"\n",
    "    if x < 0 or x > 1 : \n",
    "        # print(\"bad x in routine betai\")\n",
    "        return False\n",
    "    if x == 0 or x == 1 : \n",
    "        bt = 0.0\n",
    "    else : \n",
    "        bt = exp(lgamma(a+b)-lgamma(a)-lgamma(b)+a*log(x)+b*log(1.0-x))\n",
    "    if x < (a+1) / (a+b+2) : \n",
    "        return bt * betacf(a,b,x)/a\n",
    "    return 1 - bt*betacf(b,a,1-x)/b\n",
    "\n",
    "def get_binom_pval(n:int,k:int,p:float) -> float:\n",
    "    \"\"\"\n",
    "    This function allows to calculate the binomial probability \n",
    "    of obtaining k in a set of size n and whose probability is p \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n : int\n",
    "        Number of genomic region in the test set \n",
    "    k : int \n",
    "        Number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "    p : float\n",
    "        Percentage of genome annotated\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        binomial probability\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> get_binom_pval(100,2,0.2)\n",
    "        0.9999999947037065\n",
    "    \n",
    "    \"\"\"\n",
    "    if k == 0 : return 1\n",
    "    else : return betai(k,n-k+1,p)\n",
    "\n",
    "def hypergeom_pmf(N, K, n, k):\n",
    "    \"\"\"\n",
    "    Function to calculate the probability mass function for hypergeometric distribution\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    N : int\n",
    "        Total number of gene in the genome\n",
    "    K : int \n",
    "        Number of genes in the genome with annotation\n",
    "    n : int\n",
    "        Number of gene in the test set\n",
    "    k : int\n",
    "        Number of genes in the test gene set with annotation\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        proability mass function\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> hypergeom_pmf(100,10,30,1)\n",
    "        0.11270773995748315\n",
    "    \n",
    "    \"\"\"\n",
    "    Achoosex = comb(K,k,exact=True) \n",
    "    NAchoosenx = comb(N-K, n-k,exact=True) \n",
    "    Nchoosen = comb(N,n,exact=True) \n",
    "    return ((Achoosex)*NAchoosenx)/Nchoosen \n",
    "\n",
    "def hypergeom_cdf(N, K, n, k):\n",
    "    \"\"\"\n",
    "    Function to calculate the cumulative density funtion for hypergeometric distribution\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    N : int\n",
    "        Total number of gene in the genome\n",
    "    K : int \n",
    "        Number of genes in the genome with annotation\n",
    "    n : int\n",
    "        Number of gene in the test set\n",
    "    k : int\n",
    "        Number of genes in the test gene set with annotation\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        Cumulative density function\n",
    "        \n",
    "    Exemples \n",
    "    --------\n",
    "    >>> hypergeom_cdf(100,10,30,1)\n",
    "        0.9770827595419788\n",
    "    \n",
    "    \"\"\"\n",
    "    return np.sum([hypergeom_pmf(N, K, n, x) for x in range(k,min(K,n)+1)])\n",
    "\n",
    "class GREAT: \n",
    "    def loader(test_data:None or str or pd.DataFrame,regdom_file:None or str or pd.DataFrame,chr_size_file:None or str or pd.DataFrame,annotation_file:None or str or pd.DataFrame):\n",
    "        \"\"\"\n",
    "        This function is used to load all datasets needed for the enrichment calculation\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test_data : str or pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom_file : str or pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size_file : str or pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation_file : str or pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        test_data : pd.DataFrame\n",
    "            Genomic set of peaks to be tested in the good format \n",
    "        regdom : pd.DataFrame\n",
    "            Regulatory domain of all genes in the genome in the good format\n",
    "        size : pd.DataFrame\n",
    "            Table with the size of each chromosome in the good format\n",
    "        ann : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome in the good format\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\n",
    "        ...    \"../../data/tests/test_data/input/02_srf_hg38.bed\",\n",
    "        ...    \"../../data/human/hg38/regulatory_domain.bed\",\n",
    "        ...    \"../../data/human/hg38/chr_size.bed\",\n",
    "        ...    \"../../data/human/ontologies.csv\"\n",
    "        ...    )\n",
    "\n",
    "        >>> test.head()\n",
    "            |    | Chr   |   Chr_Start |   Chr_End |\n",
    "            |---:|:------|------------:|----------:|\n",
    "            |  0 | chr1  |     1052028 |   1052049 |\n",
    "            |  1 | chr1  |     1065512 |   1065533 |\n",
    "            |  2 | chr1  |     1067375 |   1067397 |\n",
    "            |  3 | chr1  |     1068083 |   1068119 |\n",
    "            |  4 | chr1  |    10520283 |  10520490 |\n",
    "\n",
    "        >>> regdom.head()\n",
    "            |    | Chr   |   Chr_Start |   Chr_End | Name      |   tss | Strand   |\n",
    "            |---:|:------|------------:|----------:|:----------|------:|:---------|\n",
    "            |  0 | chr1  |           0 |     22436 | MIR6859-1 | 17436 | -        |\n",
    "            |  1 | chr1  |       16436 |     22436 | MIR6859-2 | 17436 | -        |\n",
    "            |  2 | chr1  |       16436 |     22436 | MIR6859-3 | 17436 | -        |\n",
    "            |  3 | chr1  |       16436 |     28370 | MIR6859-4 | 17436 | -        |\n",
    "            |  4 | chr1  |       22436 |     34370 | WASH7P    | 29370 | -        |\n",
    "\n",
    "        >>> size.head()\n",
    "            |    | Chrom   |      Size |\n",
    "            |---:|:--------|----------:|\n",
    "            |  0 | chr1    | 248956422 |\n",
    "            |  1 | chr2    | 242193529 |\n",
    "            |  2 | chr3    | 198295559 |\n",
    "            |  3 | chr4    | 190214555 |\n",
    "            |  4 | chr5    | 181538259 |\n",
    "            \n",
    "        >>> ann.head()\n",
    "            |    | id         | name                                                   | symbol        |\n",
    "            |---:|:-----------|:-------------------------------------------------------|:--------------|\n",
    "            |  0 | GO:0003924 | GTPase activity                                        | DNAJC25-GNG10 |\n",
    "            |  1 | GO:0007186 | G protein-coupled receptor signaling pathway           | DNAJC25-GNG10 |\n",
    "            |  2 | GO:0003723 | RNA binding                                            | NUDT4B        |\n",
    "            |  3 | GO:0005829 | cytosol                                                | NUDT4B        |\n",
    "            |  4 | GO:0008486 | diphosphoinositol-polyphosphate diphosphatase activity | NUDT4B        |\n",
    "        \n",
    "        \"\"\"\n",
    "\n",
    "        if type(regdom_file) == str :\n",
    "            regdom = pd.read_csv(regdom_file,sep = \"\\t\",comment=\"#\",\n",
    "                        names = [\"Chr\", \"Chr_Start\", \"Chr_End\",\"Name\",\"tss\",\"Strand\"],\n",
    "                        dtype = {\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\",\"Name\":\"object\",\"tss\":\"int64\",\"Strand\":\"object\"})\n",
    "        elif type(regdom_file) == pd.DataFrame :\n",
    "            regdom = regdom_file.iloc[:,:6]\n",
    "            colname = list(regdom.columns)\n",
    "            try : \n",
    "                regdom = regdom.rename(columns = {colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\",colname[3]:\"Name\",colname[4]:\"tss\",colname[5]:\"Strand\"})\n",
    "            except :\n",
    "                print(\"Error in the format of the regdom file\")\n",
    "                print(\"The regdom file must have the following columns : Chr, Chr_Start, Chr_End, Name, tss, Strand\")\n",
    "                return False \n",
    "        else : \n",
    "            regdom = regdom_file\n",
    "\n",
    "        if type(test_data) == str : \n",
    "            test_data = pd.read_csv(test_data,sep = \"\\t\",comment = \"#\",usecols = [0,1,2],\n",
    "                            names = [\"Chr\", \"Chr_Start\", \"Chr_End\"],\n",
    "                            dtype = {\"Chr\":\"object\", \"Chr_Start\":\"int64\", \"Chr_End\":\"int64\"})\n",
    "        elif type(test_data) == pd.DataFrame : \n",
    "            test_data = test_data.iloc[:,:3]\n",
    "            colname = list(test_data.columns)\n",
    "            try : \n",
    "                test_data = test_data.rename(columns={colname[0]:\"Chr\",colname[1]:\"Chr_Start\",colname[2]:\"Chr_End\"})\n",
    "            except : \n",
    "                print(\"Error in test dataframe, please check your input\")\n",
    "                print(\"Columns should be : chr...(type object), start(type int), end(type int)\")\n",
    "                return False\n",
    "        else :\n",
    "            pass\n",
    "\n",
    "        if type(chr_size_file) == str :\n",
    "            size = pd.read_csv(chr_size_file,sep = \"\\t\",comment = \"#\",\n",
    "                            names = [\"Chrom\",\"Size\"],\n",
    "                            dtype = {\"Chrom\":\"object\", \"Size\":\"int64\"})\n",
    "        elif type(chr_size_file) == pd.DataFrame :\n",
    "            size = chr_size_file.iloc[:,:2]\n",
    "            colname = list(size.columns)\n",
    "            try : \n",
    "                size = size.rename(columns = {colname[0]:\"Chrom\",colname[1]:\"Size\"})\n",
    "            except : \n",
    "                print(\"Error in the format of the chr_size file\")\n",
    "                print(\"The chr_size file must have the following columns : Chrom, Size\")\n",
    "                return False\n",
    "        else :\n",
    "            size = chr_size_file\n",
    "\n",
    "        if type(annotation_file) == str : \n",
    "            dask_df = dd.read_csv(annotation_file,sep = \";\",  comment = \"#\",\n",
    "                            dtype = {\"ensembl\":\"object\",\"id\":\"object\",\"name\":\"object\",\"ontology.group\":\"object\",\"gene.name\":\"object\",\"symbol\":\"object\"},\n",
    "                            usecols = [\"id\",\"name\",\"symbol\"],low_memory = False)\n",
    "            ann = dask_df.compute()\n",
    "            ann = ann[ann['id'].str.match('^GO.*')== True]\n",
    "        elif type(annotation_file) == pd.DataFrame : \n",
    "            ann = annotation_file.iloc[:,:4]\n",
    "            colname = list(ann.columns)\n",
    "            try : \n",
    "                ann = ann.rename(columns={colname[0]:\"id\",colname[1]:\"name\",colname[3]:\"symbol\"})\n",
    "            except : \n",
    "                print(\"Error in the format of the annotation file\")\n",
    "                print(\"The annotation file must have the following columns : id, name, symbol\")\n",
    "                return False\n",
    "        else :\n",
    "            ann = annotation_file\n",
    "\n",
    "        return test_data,regdom,size,ann\n",
    "\n",
    "    def __enrichment_binom_and_hypergeom(test,regdom,size,ann,asso) : \n",
    "        \"\"\"\n",
    "        This private function is used to compute the enrichment of the test data using the binomial test and the hypergeometric test.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom : pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size :  pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        asso : list \n",
    "            List of the association between gene from regdom and peaks from test\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the binomial test and the the hypergeometric test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/tests/test_data/input/02_srf_hg38.bed\",\"../../data/human/hg38/regulatory_domain.bed\",\"../../data/human/hg38/chr_size.bed\",\"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.____enrichment_binom_and_hypergeom(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    size = size,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom)\n",
    "        ...    )\n",
    "\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  |\n",
    "        \"\"\"\n",
    "        # Init Great\n",
    "        res = {}\n",
    "        hit = {}\n",
    "\n",
    "        # init Hypergeom\n",
    "        hypergeom_gene_set = len(asso) # get the number of genes in the test gene set.\n",
    "        hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome.\n",
    "\n",
    "        # Init binom \n",
    "        n_binom = test.shape[0]# get the number of genomic region in the test set\n",
    "        total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
    "\n",
    "        ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
    "        regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
    "        len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
    "\n",
    "        #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
    "        for name in asso :\n",
    "            ann_name_gene = ann[ann[\"symbol\"].isin([name])]\n",
    "            id = ann_name_gene[\"id\"]\n",
    "            tmp = []\n",
    "            for i in (list(id.unique())) : \n",
    "                gene_imply = ann[ann['id'].isin([i])]\n",
    "                K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
    "                curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
    "                k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation\n",
    "\n",
    "                if i not in list(hit.keys()) : \n",
    "                    hit[i] = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "                k_binom = hit[i]\n",
    "                nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
    "                tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom))\n",
    "            res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu),elem[0]/(elem[1]/total_nu),n_binom,elem[0],elem[1]/total_nu, hypergeom_cdf(hypergeom_total_number_gene,elem[4],hypergeom_gene_set,elem[5]),(elem[5]*hypergeom_total_number_gene)/(hypergeom_gene_set*elem[4]),hypergeom_total_number_gene,elem[4],hypergeom_gene_set,elem[5] ] for elem in tmp})\n",
    "        return pd.DataFrame(res).transpose().rename(columns = {0:\"go_term\",1:\"binom_p_value\",2:\"binom_fold_enrichment\",3:\"n_binom\",4:\"k_binom\",5:\"p_binom\",6:\"hypergeom_p_value\",7:\"hypergeometric_fold_enrichment\",8:\"N_hyper\",9:\"K_hyper\",10:\"n_hyper\",11:\"k_hyper\"}).replace(0,np.nan).dropna().sort_values(by = \"binom_p_value\")\n",
    "    \n",
    "    def __enrichment_binom(test,regdom,size,ann,asso):\n",
    "        \"\"\"\n",
    "        This private function is used to compute the enrichment of the test data using the binomial test.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom : pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size :  pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        asso : list \n",
    "            List of the association between gene from regdom and peaks from test\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the binomial test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/tests/test_data/input/02_srf_hg38.bed\",\"../../data/human/hg38/regulatory_domain.bed\",\"../../data/human/hg38/chr_size.bed\",\"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.____enrichment_binom(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    size = size,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom)\n",
    "        ...    )\n",
    "\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |\n",
    "        \"\"\" \n",
    "        # Init Great\n",
    "        res = {}\n",
    "        hit = {}\n",
    "\n",
    "        # Init binom \n",
    "        n_binom = test.shape[0]# get the number of genomic region in the test set\n",
    "        total_nu = size[\"Size\"].sum()# get the total number of nucleotides in the genome\n",
    "        \n",
    "        ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
    "        regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
    "        len_on_chr = len_regdom(regdom)# get the length of each regulatory domain \n",
    "\n",
    "        #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
    "        for name in asso :\n",
    "            ann_name_gene = ann[ann[\"symbol\"].isin([name])]\n",
    "            id = ann_name_gene[\"id\"]\n",
    "            tmp = []\n",
    "            for i in (list(id.unique())) : \n",
    "                gene_imply = ann[ann['id'].isin([i])]\n",
    "                curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
    "\n",
    "                if i not in list(hit.keys()) : \n",
    "                    hit[i] = number_of_hit(test,curr_regdom)# get the number of test genomic regions in the regulatory domain of a gene with annotation\n",
    "                k_binom = hit[i]\n",
    "                nb_binom = sum([len_on_chr[i] for i in curr_regdom[\"Name\"]])# get the portion of the genome in the regulatory domain of a gene with annotation\n",
    "                tmp.append((k_binom,nb_binom,i,gene_imply.iloc[0][\"name\"]))\n",
    "            res.update({elem[2]:[ elem[3],get_binom_pval(n_binom,elem[0],elem[1]/total_nu),elem[0]/(elem[1]/total_nu) ] for elem in tmp})\n",
    "        return pd.DataFrame(res).transpose().rename(columns = {0:\"go_term\",1:\"binom_p_value\",2:\"binom_fold_enrichment\"}).sort_values(by = \"binom_p_value\")\n",
    "\n",
    "    def __enrichment_hypergeom(test,regdom,ann,asso): \n",
    "        \"\"\"\n",
    "        This private function is used to compute the enrichment of the test data using the hypergeometric test.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom : pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size :  pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        asso : list \n",
    "            List of the association between gene from regdom and peaks from test\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the hypergeometric test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.____enrichment_hypergeom(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom)\n",
    "        ...    )\n",
    "\n",
    "        >>> enrichment.head()   \n",
    "        \"\"\" \n",
    "        # Init Great\n",
    "        res = {}\n",
    "\n",
    "        # Init hypergeom\n",
    "        hypergeom_total_number_gene = regdom.shape[0] #get the number of genes in the genome\n",
    "        hypergeom_gene_set = len(asso) # get the number of genes in the test gene set.\n",
    "\n",
    "        ann_red = ann[ann[\"symbol\"].isin(asso)]\n",
    "        regdom = regdom[regdom[\"Name\"].isin(list(ann[ann[\"id\"].isin(list(ann_red[\"id\"]))][\"symbol\"]))]#reduction of the regdom file by selecting only the genes whose GO ID is owned by a gene of the association \n",
    "\n",
    "        #Compute for all associating gene and for each GO id associated with the gene the probability. \n",
    "        for name in asso :\n",
    "            ann_name_gene = ann[ann[\"symbol\"] == name]\n",
    "            id = ann_name_gene[\"id\"]\n",
    "            tmp = []\n",
    "            for i in (list(id.unique())) : \n",
    "                gene_imply = ann[ann['id']==i]\n",
    "                K_hypergeom = gene_imply.shape[0] # get be the number of genes in the genome with annotation\n",
    "                curr_regdom = regdom.loc[regdom[\"Name\"].isin(list(gene_imply[\"symbol\"]))]\n",
    "                k_hypergeom = curr_regdom.loc[curr_regdom[\"Name\"].isin(asso)].shape[0] # get the number of genes in the test gene set with annotation                \n",
    "                tmp.append((i,gene_imply.iloc[0][\"name\"],K_hypergeom,k_hypergeom)) \n",
    "            res.update({elem[0]:[ elem[1], hypergeom_cdf(hypergeom_total_number_gene,elem[2],hypergeom_gene_set,elem[3]),(elem[3]*hypergeom_total_number_gene)/(hypergeom_gene_set*elem[2]) ] for elem in tmp}) \n",
    "        return pd.DataFrame(res).transpose().rename(columns = {0:\"go_term\",1:\"hypergeom_p_value\",2:\"hypergeometric_fold_enrichment\"}).replace(0,np.nan).dropna().sort_values(by = \"hypergeom_p_value\")\n",
    "\n",
    "\n",
    "    def enrichment(test_file,regdom_file,chr_size_file, annotation_file, binom=True,hypergeom=True):\n",
    "        \"\"\"\n",
    "        This function is a wrapper of the 3 private methods: \n",
    "        * GREAT.__enrichment_binom_and_hypergeom \n",
    "        * GREAT.__enrichment_binom \n",
    "        * GREAT.__enrichment_hypergeom\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test_file : pd.DataFrame\n",
    "            Genomic set of peaks to be tested\n",
    "        regdom_file : pd.DataFrame \n",
    "            Regulatory domain of all genes in the genome \n",
    "        chr_size_file :  pd.DataFrame\n",
    "            Table with the size of each chromosome\n",
    "        annotation_file : pd.DataFrame\n",
    "            Table with the annotation of each gene in the genome\n",
    "        binom : bool\n",
    "            If True, the binomial test is used.\n",
    "        hypergeom : bool\n",
    "            If True, the hypergeometric test is used.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe contains for every GO ID associate with a every associated gene the p-value for the hypergeometric test\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = GREAT.enrichment(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    chr_size_file = size,\n",
    "        ...    ann = ann,\n",
    "        ...    binom=True,\n",
    "        ...    hypergeom=True\n",
    "        ...    )\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  |\n",
    "\n",
    "        >>> enrichment = GREAT.enrichment(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom),\n",
    "        ...    binom=True,\n",
    "        ...    hypergeom=False\n",
    "        ...    )\n",
    "        >>> enrichment.head()\n",
    "            |            | go_term                                                          |   binom_p_value |\n",
    "            |:-----------|:-----------------------------------------------------------------|----------------:|\n",
    "            | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |\n",
    "            | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |\n",
    "            | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |\n",
    "            | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |\n",
    "            | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |\n",
    "\n",
    "        >>> enrichment = GREAT.enrichment(\n",
    "        ...    test = test,\n",
    "        ...    regdom = regdom,\n",
    "        ...    ann = ann,\n",
    "        ...    asso = get_association(test,regdom),\n",
    "        ...    binom=False,\n",
    "        ...    hypergeom=True\n",
    "        ...    )\n",
    "        >>> enrichment.head()\n",
    "\n",
    "        \"\"\"\n",
    "        if not binom and not hypergeom : \n",
    "            return False\n",
    "        \n",
    "        test,regdom,size,ann = GREAT.loader(test_file,regdom_file,chr_size_file, annotation_file)\n",
    "        asso = get_association(test,regdom)# get the name of the regulatory domain associated to each genomic region in the test set\n",
    "\n",
    "\n",
    "        if binom and hypergeom : \n",
    "            return GREAT.__enrichment_binom_and_hypergeom(test,regdom,size,ann,asso)\n",
    "\n",
    "        elif binom : \n",
    "            return GREAT.__enrichment_binom(test,regdom,size,ann,asso)\n",
    "\n",
    "        else : \n",
    "              return GREAT.__enrichment_hypergeom(test,regdom,ann,asso)\n",
    "\n",
    "    def set_bonferroni(self,alpha:float=0.05): \n",
    "        \"\"\"\n",
    "        This function create new columns in the dataframe with the Bonferroni correction\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        alpha : float\n",
    "            alpha value for the Bonferroni correction\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe new columns with the Bonferroni correction for each p-value\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = great.tl.GREAT.enrichment(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\",binom=True,hypergeom=True)\n",
    "        >>> bonferroni = GREAT.set_bonferroni(enrichment,alpha=0.05)\n",
    "        >>> bonferroni.head()\n",
    "        ...    |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |   binom_bonferroni |   hypergeom_bonferroni |\n",
    "        ...    |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|-------------------:|-----------------------:|\n",
    "        ...    | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  |        3.0754e-10  |                      1 |\n",
    "        ...    | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  |        2.87384e-07 |                      1 |\n",
    "        ...    | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  |        2.87384e-07 |                      1 |\n",
    "        ...    | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 |        3.74892e-07 |                      1 |\n",
    "        ...    | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  |        7.68931e-07 |                      1 |\n",
    "\n",
    "        \"\"\"\n",
    "        for col in self.columns : \n",
    "            if col in [\"binom_p_value\",\"hypergeom_p_value\"] : \n",
    "                col_split = col.split(\"_\")\n",
    "                self[f\"{col_split[0]}_bonferroni\"] = multipletests(self[col], alpha=alpha, method='bonferroni')[1]\n",
    "        return self \n",
    "\n",
    "    def set_fdr(self,alpha:float=0.05) : \n",
    "        \"\"\"\n",
    "        This function create new columns in the dataframe with the fdr correction\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        alpha : float\n",
    "            alpha value for the fdr correction\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe new columns with the fdr correction for each p-value\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = great.tl.GREAT.enrichment(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\",binom=True,hypergeom=True)\n",
    "        >>> fdr = GREAT.set_fdr(enrichment,alpha=0.05)\n",
    "        >>> fdr.head()\n",
    "        ...    |            | go_term                                                          |   binom_p_value |   hypergeom_p_value |   binom_fdr |   hypergeom_fdr |\n",
    "        ...    |:-----------|:-----------------------------------------------------------------|----------------:|--------------------:|------------:|----------------:|\n",
    "        ...    | GO:0045887 | positive regulation of synaptic growth at neuromuscular junction |     5.17744e-13 |          0.0029275  | 3.0754e-10  |       0.0913909 |\n",
    "        ...    | GO:0044721 | protein import into peroxisome matrix, substrate release         |     4.83812e-10 |          0.0029275  | 9.3723e-08  |       0.0913909 |\n",
    "        ...    | GO:0036250 | peroxisome transport along microtubule                           |     4.83812e-10 |          0.0029275  | 9.3723e-08  |       0.0913909 |\n",
    "        ...    | GO:0016561 | protein import into peroxisome matrix, translocation             |     6.31131e-10 |          0.00584656 | 9.3723e-08  |       0.0913909 |\n",
    "        ...    | GO:0047485 | protein N-terminus binding                                       |     1.2945e-09  |          0.0050377  | 1.53786e-07 |       0.0913909 |\n",
    "\n",
    "        \"\"\"\n",
    "        for col in self.columns : \n",
    "            if col in [\"binom_p_value\",\"hypergeom_p_value\"] : \n",
    "                col_split = col.split(\"_\")\n",
    "                # self[f\"{col_split[0]}_fdr\"] = fdrcorrection(self[col], alpha=alpha)[1]\n",
    "                self[f\"{col_split[0]}_fdr\"] = multipletests(self[col], alpha=alpha, method='fdr_bh')[1]\n",
    "        return self \n",
    "\n",
    "    def set_threshold(self,colname:str, alpha:int=0.05) : \n",
    "        \"\"\"\n",
    "        This function allows to delete rows according to the p-value of the column taken as argument. By default the alpha value is 0.05\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        alpha : float\n",
    "            alpha value for the fdr correction\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "            dataframe with the rows deleted according to the p-value threshold\n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        >>> test,regdom,size,ann = GREAT.loader(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\")\n",
    "        >>> enrichment = great.tl.GREAT.enrichment(\"../../data/human/test_genomic_region.bed\", \"../../data/human/regulatory_domain.bed\", \"../../data/human/chr_size.bed\", \"../../data/human/ontologies.csv\",binom=True,hypergeom=True)\n",
    "        >>> enrichment.shape[0]\n",
    "        ...    594\n",
    "\n",
    "        >>> significant = GREAT.set_threshold(enrichment,colname=\"binom_p_value\",alpha=0.05)\n",
    "        >>> significant.shape[0]\n",
    "        ...    310\n",
    "\n",
    "        \"\"\"\n",
    "        if colname in self.columns: \n",
    "            self = self.loc[self[colname]<=alpha]\n",
    "        return self \n",
    "\n",
    "###################################################################\n",
    "########################### unfinish ##############################\n",
    "###################################################################\n",
    "\n",
    "    def get_nb_asso_per_region(test,regdom) : \n",
    "        \"\"\"\n",
    "        Function allowing from a file of genomic regions from CHIPseq \n",
    "        and a file of genomic regulatory domains to determine number of peaks \n",
    "        associated with each gene in the regulatory domain. \n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : str\n",
    "            path of the file with the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "        \n",
    "        regdom : str\n",
    "            path of the file with the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        res : dict\n",
    "            dict with the number of associated genes per genomic region : key = associated gene, value = number of peaks associated with the gene \n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        test = pd.DataFrame({\n",
    "        ...    \"Chr\":[\"chr1\"],\n",
    "        ...    \"Chr_Start\":[1052028],\n",
    "        ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "        regdom = pd.DataFrame({\n",
    "        ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "        ...    \"Chr_Start\":[1034992,1079306],\n",
    "        ...    \"Chr_End\": [1115089,1132016],\n",
    "        ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "        ...    \"tss\":[1074306,1116089],\n",
    "        ...    \"strand\":['-','-']})\n",
    "\n",
    "        >>> get_association(test,regdom)        \n",
    "            {'RNF223':2}\n",
    "        \n",
    "        \"\"\"\n",
    "        res = {}\n",
    "        test,regdom,_,_ = GREAT.loader(test,regdom,None,None)\n",
    "        for i in range(test.shape[0]) :\n",
    "            currTest = test.iloc[i]\n",
    "            regdom_curr_test = regdom.loc[regdom[\"Chr\"] == currTest[\"Chr\"]].sort_values(\"Chr_Start\")\n",
    "            regdom_inf = regdom_curr_test.loc[regdom_curr_test[\"tss\"] <= currTest[\"Chr_Start\"]]\n",
    "            regdom_sup = regdom_curr_test.loc[regdom_curr_test[\"tss\"] >= currTest[\"Chr_End\"]]\n",
    "            try : \n",
    "                if regdom_inf.iloc[-1][\"Name\"] not in res.keys() : \n",
    "                    res[regdom_inf.iloc[-1][\"Name\"]] = 1 \n",
    "                else :\n",
    "                    res[regdom_inf.iloc[-1][\"Name\"]] += 1\n",
    "            except :\n",
    "                pass\n",
    "\n",
    "            try :\n",
    "                if regdom_sup.iloc[-1][\"Name\"] not in res.keys() : \n",
    "                    res[regdom_sup.iloc[-1][\"Name\"]] = 1 \n",
    "                else :\n",
    "                    res[regdom_sup.iloc[-1][\"Name\"]] += 1\n",
    "            except : \n",
    "                pass\n",
    "        return res\n",
    "\n",
    "    def get_dist_to_tss(test,regdom) : \n",
    "        \"\"\"\n",
    "        Function allowing from a file of genomic regions from CHIPseq \n",
    "        and a file of genomic regulatory domains to determine the distance from peaks \n",
    "        to the transcription start site of the associated gene\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        test : str\n",
    "            path of the file with the tests pics => columns: [\"Chr\",\"Chr_Start\",\"Chr_End\"]\n",
    "        \n",
    "        regdom : str\n",
    "            path of the file with the regulatory domains => columns: [\"Chr\"\t\"Chr_Start\"\t\"Chr_End\"\t\"Name\"\t\"tss\"\t\"strand\"].\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        res : dict\n",
    "            dict with the distance from tss to the associated genes : key = associated gene, value = distance from peaks to tss \n",
    "            \n",
    "        Exemples \n",
    "        --------\n",
    "        test = pd.DataFrame({\n",
    "        ...    \"Chr\":[\"chr1\"],\n",
    "        ...    \"Chr_Start\":[1052028],\n",
    "        ...    \"Chr_End\": [1052049]})\n",
    "\n",
    "        regdom = pd.DataFrame({\n",
    "        ...    \"Chr\":[\"chr1\",\"chr1\"],\n",
    "        ...    \"Chr_Start\":[1034992,1079306],\n",
    "        ...    \"Chr_End\": [1115089,1132016],\n",
    "        ...    \"Name\":[\"RNF223\",\"C1orf159\"],\n",
    "        ...    \"tss\":[1074306,1116089],\n",
    "        ...    \"strand\":['-','-']})\n",
    "\n",
    "        >>> get_association(test,regdom)        \n",
    "            {'RNF223':[-22278]}\n",
    "        \n",
    "        \"\"\"\n",
    "        res = {}\n",
    "        test,regdom,_,_ = GREAT.loader(test,regdom,None,None)\n",
    "        for i in range(test.shape[0]) :\n",
    "            currTest = test.iloc[i]\n",
    "            regdom_curr_test = regdom.loc[regdom[\"Chr\"] == currTest[\"Chr\"]].sort_values(\"Chr_Start\")\n",
    "            regdom_inf = regdom_curr_test.loc[regdom_curr_test[\"tss\"] <= currTest[\"Chr_Start\"]]\n",
    "            regdom_sup = regdom_curr_test.loc[regdom_curr_test[\"tss\"] >= currTest[\"Chr_End\"]]\n",
    "            try : \n",
    "                if regdom_inf.iloc[-1][\"Name\"] not in res.keys() : \n",
    "                    res[regdom_inf.iloc[-1][\"Name\"]] = [currTest[\"Chr_Start\"] - res[regdom_inf.iloc[-1][\"tss\"]]]\n",
    "                else : \n",
    "                    res[regdom_inf.iloc[-1][\"Name\"]].append(currTest[\"Chr_Start\"] - res[regdom_inf.iloc[-1][\"tss\"]])\n",
    "            except :\n",
    "                pass\n",
    "\n",
    "            try :\n",
    "                if regdom_sup.iloc[-1][\"Name\"] not in res.keys() : \n",
    "                    res[regdom_sup.iloc[-1][\"Name\"]] = [currTest[\"Chr_End\"] - res[regdom_sup.iloc[-1][\"tss\"]]]\n",
    "                else :\n",
    "                    res[regdom_sup.iloc[-1][\"Name\"]].append(currTest[\"Chr_End\"] - res[regdom_sup.iloc[-1][\"tss\"]])\n",
    "            except : \n",
    "                pass\n",
    "        return res\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "rgreat = importr('rGREAT')\n",
    "ranges = importr('GenomicRanges')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = \"../../data/tests/test_data/input/01_random.bed\"\n",
    "regdom = \"../../data/human/hg38/regulatory_domain.bed\"\n",
    "great_out = \"../../data/tests/test_data/output/10_MAX_output_great_webserver.tsv\"\n",
    "great_asso = \"../../data/tests/test_data/output/10_MAX_gene_asso.tsv\"\n",
    "size = \"../../data/human/hg38/chr_size.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: * use genome 'hg38'.\n",
      "\n",
      "R[write to console]: * TSS extension mode is 'basalPlusExt'.\n",
      "\n",
      "R[write to console]: * construct the basal domains by extending 5000bp to upstream and 1000bp to downsteram of TSS.\n",
      "\n",
      "R[write to console]: * calculate distances to neighbour regions.\n",
      "\n",
      "R[write to console]: * extend to both sides until reaching the neighbour genes or to the maximal extension (1000000bp).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df = r[\"read.csv\"](test,sep='\\t')\n",
    "seqname = rpy2.robjects.StrVector([\"seqnames\", \"seqname\",\"chromosome\", \"X.Chr\",\"chr\", \"chromosome_name\",\"seqid\"])\n",
    "end = rpy2.robjects.StrVector(['end', 'stop'])\n",
    "df = ranges.makeGRangesFromDataFrame(df,seqnames_field=seqname)#,keep_extra_columns=False,ignore_strand=False,seqinfo=None,start_field='start', end_field=end, strand_field='strand',starts_in_df_are_0based=False)\n",
    "\n",
    "    # great calculation\n",
    "local = rgreat.great(df, \"msigdb:C5\", f\"GREAT:hg38\",verbose=False)\n",
    "local = rgreat.getEnrichmentTables(local)\n",
    "\n",
    "time.sleep(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# greatpy calculation\n",
    "greatpy = GREAT.enrichment(\n",
    "    test_file=test,\n",
    "    regdom_file=regdom,\n",
    "    chr_size_file=size,\n",
    "    annotation_file=f\"../../data/human/ontologies.csv\",\n",
    "    binom=True,\n",
    "    hypergeom=True,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go_term</th>\n",
       "      <th>binom_p_value</th>\n",
       "      <th>binom_fold_enrichment</th>\n",
       "      <th>n_binom</th>\n",
       "      <th>k_binom</th>\n",
       "      <th>p_binom</th>\n",
       "      <th>hypergeom_p_value</th>\n",
       "      <th>hypergeometric_fold_enrichment</th>\n",
       "      <th>N_hyper</th>\n",
       "      <th>K_hyper</th>\n",
       "      <th>n_hyper</th>\n",
       "      <th>k_hyper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0045887</th>\n",
       "      <td>positive regulation of synaptic growth at neur...</td>\n",
       "      <td>5.20389e-13</td>\n",
       "      <td>2.18004e+05</td>\n",
       "      <td>104</td>\n",
       "      <td>4</td>\n",
       "      <td>1.83483e-05</td>\n",
       "      <td>3.08889e-03</td>\n",
       "      <td>3.23741e+02</td>\n",
       "      <td>18777</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0044721</th>\n",
       "      <td>protein import into peroxisome matrix, substra...</td>\n",
       "      <td>4.84198e-10</td>\n",
       "      <td>3.94065e+04</td>\n",
       "      <td>104</td>\n",
       "      <td>4</td>\n",
       "      <td>1.01506e-04</td>\n",
       "      <td>3.08889e-03</td>\n",
       "      <td>3.23741e+02</td>\n",
       "      <td>18777</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0036250</th>\n",
       "      <td>peroxisome transport along microtubule</td>\n",
       "      <td>4.84198e-10</td>\n",
       "      <td>3.94065e+04</td>\n",
       "      <td>104</td>\n",
       "      <td>4</td>\n",
       "      <td>1.01506e-04</td>\n",
       "      <td>3.08889e-03</td>\n",
       "      <td>3.23741e+02</td>\n",
       "      <td>18777</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0016561</th>\n",
       "      <td>protein import into peroxisome matrix, translo...</td>\n",
       "      <td>6.31855e-10</td>\n",
       "      <td>3.68645e+04</td>\n",
       "      <td>104</td>\n",
       "      <td>4</td>\n",
       "      <td>1.08505e-04</td>\n",
       "      <td>6.16839e-03</td>\n",
       "      <td>1.61871e+02</td>\n",
       "      <td>18777</td>\n",
       "      <td>2</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0030576</th>\n",
       "      <td>Cajal body organization</td>\n",
       "      <td>1.49190e-09</td>\n",
       "      <td>2.97236e+04</td>\n",
       "      <td>104</td>\n",
       "      <td>4</td>\n",
       "      <td>1.34573e-04</td>\n",
       "      <td>1.83932e-02</td>\n",
       "      <td>5.39569e+01</td>\n",
       "      <td>18777</td>\n",
       "      <td>6</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006355</th>\n",
       "      <td>regulation of transcription, DNA-templated</td>\n",
       "      <td>9.93956e-01</td>\n",
       "      <td>2.08617e+01</td>\n",
       "      <td>104</td>\n",
       "      <td>1</td>\n",
       "      <td>4.79347e-02</td>\n",
       "      <td>8.40449e-01</td>\n",
       "      <td>5.54352e-01</td>\n",
       "      <td>18777</td>\n",
       "      <td>584</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006915</th>\n",
       "      <td>apoptotic process</td>\n",
       "      <td>9.95014e-01</td>\n",
       "      <td>2.01230e+01</td>\n",
       "      <td>104</td>\n",
       "      <td>1</td>\n",
       "      <td>4.96944e-02</td>\n",
       "      <td>5.39651e-01</td>\n",
       "      <td>1.11443e+00</td>\n",
       "      <td>18777</td>\n",
       "      <td>581</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042803</th>\n",
       "      <td>protein homodimerization activity</td>\n",
       "      <td>9.98551e-01</td>\n",
       "      <td>2.44788e+01</td>\n",
       "      <td>104</td>\n",
       "      <td>2</td>\n",
       "      <td>8.17033e-02</td>\n",
       "      <td>9.54394e-01</td>\n",
       "      <td>3.33067e-01</td>\n",
       "      <td>18777</td>\n",
       "      <td>972</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005737</th>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>9.98903e-01</td>\n",
       "      <td>6.55886e+01</td>\n",
       "      <td>104</td>\n",
       "      <td>24</td>\n",
       "      <td>3.65917e-01</td>\n",
       "      <td>9.92060e-01</td>\n",
       "      <td>5.70469e-01</td>\n",
       "      <td>18777</td>\n",
       "      <td>5675</td>\n",
       "      <td>58</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005887</th>\n",
       "      <td>integral component of plasma membrane</td>\n",
       "      <td>9.99465e-01</td>\n",
       "      <td>3.52025e+01</td>\n",
       "      <td>104</td>\n",
       "      <td>5</td>\n",
       "      <td>1.42035e-01</td>\n",
       "      <td>9.63640e-01</td>\n",
       "      <td>4.04677e-01</td>\n",
       "      <td>18777</td>\n",
       "      <td>1600</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>579 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      go_term  binom_p_value  \\\n",
       "GO:0045887  positive regulation of synaptic growth at neur...    5.20389e-13   \n",
       "GO:0044721  protein import into peroxisome matrix, substra...    4.84198e-10   \n",
       "GO:0036250             peroxisome transport along microtubule    4.84198e-10   \n",
       "GO:0016561  protein import into peroxisome matrix, translo...    6.31855e-10   \n",
       "GO:0030576                            Cajal body organization    1.49190e-09   \n",
       "...                                                       ...            ...   \n",
       "GO:0006355         regulation of transcription, DNA-templated    9.93956e-01   \n",
       "GO:0006915                                  apoptotic process    9.95014e-01   \n",
       "GO:0042803                  protein homodimerization activity    9.98551e-01   \n",
       "GO:0005737                                          cytoplasm    9.98903e-01   \n",
       "GO:0005887              integral component of plasma membrane    9.99465e-01   \n",
       "\n",
       "            binom_fold_enrichment  n_binom  k_binom      p_binom  \\\n",
       "GO:0045887            2.18004e+05      104        4  1.83483e-05   \n",
       "GO:0044721            3.94065e+04      104        4  1.01506e-04   \n",
       "GO:0036250            3.94065e+04      104        4  1.01506e-04   \n",
       "GO:0016561            3.68645e+04      104        4  1.08505e-04   \n",
       "GO:0030576            2.97236e+04      104        4  1.34573e-04   \n",
       "...                           ...      ...      ...          ...   \n",
       "GO:0006355            2.08617e+01      104        1  4.79347e-02   \n",
       "GO:0006915            2.01230e+01      104        1  4.96944e-02   \n",
       "GO:0042803            2.44788e+01      104        2  8.17033e-02   \n",
       "GO:0005737            6.55886e+01      104       24  3.65917e-01   \n",
       "GO:0005887            3.52025e+01      104        5  1.42035e-01   \n",
       "\n",
       "            hypergeom_p_value  hypergeometric_fold_enrichment  N_hyper  \\\n",
       "GO:0045887        3.08889e-03                     3.23741e+02    18777   \n",
       "GO:0044721        3.08889e-03                     3.23741e+02    18777   \n",
       "GO:0036250        3.08889e-03                     3.23741e+02    18777   \n",
       "GO:0016561        6.16839e-03                     1.61871e+02    18777   \n",
       "GO:0030576        1.83932e-02                     5.39569e+01    18777   \n",
       "...                       ...                             ...      ...   \n",
       "GO:0006355        8.40449e-01                     5.54352e-01    18777   \n",
       "GO:0006915        5.39651e-01                     1.11443e+00    18777   \n",
       "GO:0042803        9.54394e-01                     3.33067e-01    18777   \n",
       "GO:0005737        9.92060e-01                     5.70469e-01    18777   \n",
       "GO:0005887        9.63640e-01                     4.04677e-01    18777   \n",
       "\n",
       "            K_hyper  n_hyper  k_hyper  \n",
       "GO:0045887        1       58        1  \n",
       "GO:0044721        1       58        1  \n",
       "GO:0036250        1       58        1  \n",
       "GO:0016561        2       58        1  \n",
       "GO:0030576        6       58        1  \n",
       "...             ...      ...      ...  \n",
       "GO:0006355      584       58        1  \n",
       "GO:0006915      581       58        2  \n",
       "GO:0042803      972       58        1  \n",
       "GO:0005737     5675       58       10  \n",
       "GO:0005887     1600       58        2  \n",
       "\n",
       "[579 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greatpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"id\"                    \"genome_fraction\"       \"observed_region_hits\" \n",
      " [4] \"fold_enrichment\"       \"p_value\"               \"p_adjust\"             \n",
      " [7] \"mean_tss_dist\"         \"observed_gene_hits\"    \"gene_set_size\"        \n",
      "[10] \"fold_enrichment_hyper\" \"p_value_hyper\"         \"p_adjust_hyper\"       \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(local.names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>binom_p_val</th>\n",
       "      <th>hyper_p_val</th>\n",
       "      <th>p_binom</th>\n",
       "      <th>k_binom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>retrograde vesicle mediated transport golgi to...</td>\n",
       "      <td>3.55271e-15</td>\n",
       "      <td>1.12145e-02</td>\n",
       "      <td>4.37301e-03</td>\n",
       "      <td>1.20000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>golgi vesicle transport</td>\n",
       "      <td>1.36486e-11</td>\n",
       "      <td>1.30739e-02</td>\n",
       "      <td>2.40640e-02</td>\n",
       "      <td>1.70000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>regulation of deoxyribonuclease activity</td>\n",
       "      <td>1.71753e-09</td>\n",
       "      <td>4.13466e-04</td>\n",
       "      <td>5.72306e-04</td>\n",
       "      <td>5.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>flemming body</td>\n",
       "      <td>2.95092e-08</td>\n",
       "      <td>5.20334e-03</td>\n",
       "      <td>3.67626e-03</td>\n",
       "      <td>7.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cis golgi network</td>\n",
       "      <td>3.60733e-08</td>\n",
       "      <td>4.77428e-05</td>\n",
       "      <td>5.83826e-03</td>\n",
       "      <td>8.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>regulation of intracellular signal transduction</td>\n",
       "      <td>9.96128e-01</td>\n",
       "      <td>7.61276e-01</td>\n",
       "      <td>1.65203e-01</td>\n",
       "      <td>6.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>regulation of protein modification process</td>\n",
       "      <td>9.97033e-01</td>\n",
       "      <td>9.51139e-01</td>\n",
       "      <td>1.51710e-01</td>\n",
       "      <td>5.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>synapse</td>\n",
       "      <td>9.97216e-01</td>\n",
       "      <td>9.16143e-01</td>\n",
       "      <td>1.70161e-01</td>\n",
       "      <td>6.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>cellular response to stress</td>\n",
       "      <td>9.98265e-01</td>\n",
       "      <td>9.26567e-01</td>\n",
       "      <td>1.59363e-01</td>\n",
       "      <td>5.00000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>intrinsic component of plasma membrane</td>\n",
       "      <td>9.99497e-01</td>\n",
       "      <td>9.07910e-01</td>\n",
       "      <td>1.94559e-01</td>\n",
       "      <td>6.00000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>541 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  name  binom_p_val  \\\n",
       "0    retrograde vesicle mediated transport golgi to...  3.55271e-15   \n",
       "1                              golgi vesicle transport  1.36486e-11   \n",
       "2             regulation of deoxyribonuclease activity  1.71753e-09   \n",
       "3                                        flemming body  2.95092e-08   \n",
       "4                                    cis golgi network  3.60733e-08   \n",
       "..                                                 ...          ...   \n",
       "536    regulation of intracellular signal transduction  9.96128e-01   \n",
       "537         regulation of protein modification process  9.97033e-01   \n",
       "538                                            synapse  9.97216e-01   \n",
       "539                        cellular response to stress  9.98265e-01   \n",
       "540             intrinsic component of plasma membrane  9.99497e-01   \n",
       "\n",
       "     hyper_p_val      p_binom      k_binom  \n",
       "0    1.12145e-02  4.37301e-03  1.20000e+01  \n",
       "1    1.30739e-02  2.40640e-02  1.70000e+01  \n",
       "2    4.13466e-04  5.72306e-04  5.00000e+00  \n",
       "3    5.20334e-03  3.67626e-03  7.00000e+00  \n",
       "4    4.77428e-05  5.83826e-03  8.00000e+00  \n",
       "..           ...          ...          ...  \n",
       "536  7.61276e-01  1.65203e-01  6.00000e+00  \n",
       "537  9.51139e-01  1.51710e-01  5.00000e+00  \n",
       "538  9.16143e-01  1.70161e-01  6.00000e+00  \n",
       "539  9.26567e-01  1.59363e-01  5.00000e+00  \n",
       "540  9.07910e-01  1.94559e-01  6.00000e+00  \n",
       "\n",
       "[541 rows x 5 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_local = list(local.rx2(\"id\")) \n",
    "name_local = [\" \".join(cdc.lower().split(\"_\")[1:]) for cdc in list(local.rx2(\"id\"))]\n",
    "local_df = pd.DataFrame({\n",
    "    \"name\": name_local,\n",
    "    \"binom_p_val\" : list(local.rx2(\"p_value\")),\n",
    "    \"hyper_p_val\" : list(local.rx2(\"p_value_hyper\")),\n",
    "    \"p_binom\" : list(local.rx2(\"genome_fraction\")),\n",
    "    \"k_binom\" : list(local.rx2(\"observed_region_hits\"))\n",
    "    })\n",
    "local_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "ann = pd.read_csv(\"../../data/human/ontologies.csv\",sep=';',comment='#',header=0,)\n",
    "ann[\"name\"] = ann['name'].str.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "go = []\n",
    "for i in range(local_df.shape[0]) : \n",
    "    curr = local_df.iloc[i]\n",
    "\n",
    "    if ann.loc[ann[\"name\"].isin([curr[\"name\"]])].shape[0]>0 : \n",
    "        go.append(ann.loc[ann[\"name\"].isin([curr[\"name\"]])].iloc[0][\"id\"])\n",
    "    else :\n",
    "        go.append(\"\")\n",
    "local_df[\"id\"] = go \n",
    "local_df = local_df.loc[local_df[\"id\"]!=\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "loc_gpy = local_df[local_df[\"id\"].isin(list(greatpy.index))]\n",
    "gpy_loc = greatpy[greatpy.index.isin(list(local_df[\"id\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go_term</th>\n",
       "      <th>binom_p_value</th>\n",
       "      <th>binom_fold_enrichment</th>\n",
       "      <th>n_binom</th>\n",
       "      <th>k_binom</th>\n",
       "      <th>p_binom</th>\n",
       "      <th>hypergeom_p_value</th>\n",
       "      <th>hypergeometric_fold_enrichment</th>\n",
       "      <th>N_hyper</th>\n",
       "      <th>K_hyper</th>\n",
       "      <th>n_hyper</th>\n",
       "      <th>k_hyper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0043547</th>\n",
       "      <td>positive regulation of GTPase activity</td>\n",
       "      <td>1.10493e-03</td>\n",
       "      <td>3.13301e+02</td>\n",
       "      <td>104</td>\n",
       "      <td>11</td>\n",
       "      <td>3.51100e-02</td>\n",
       "      <td>2.48282e-02</td>\n",
       "      <td>3.60715e+00</td>\n",
       "      <td>18777</td>\n",
       "      <td>359</td>\n",
       "      <td>58</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           go_term  binom_p_value  \\\n",
       "GO:0043547  positive regulation of GTPase activity    1.10493e-03   \n",
       "\n",
       "            binom_fold_enrichment  n_binom  k_binom      p_binom  \\\n",
       "GO:0043547            3.13301e+02      104       11  3.51100e-02   \n",
       "\n",
       "            hypergeom_p_value  hypergeometric_fold_enrichment  N_hyper  \\\n",
       "GO:0043547        2.48282e-02                     3.60715e+00    18777   \n",
       "\n",
       "            K_hyper  n_hyper  k_hyper  \n",
       "GO:0043547      359       58        4  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gpy_loc[gpy_loc.index == \"GO:0043547\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>binom_p_val</th>\n",
       "      <th>hyper_p_val</th>\n",
       "      <th>p_binom</th>\n",
       "      <th>k_binom</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>positive regulation of gtpase activity</td>\n",
       "      <td>2.67021e-03</td>\n",
       "      <td>5.01449e-02</td>\n",
       "      <td>2.86761e-02</td>\n",
       "      <td>8.00000e+00</td>\n",
       "      <td>GO:0043547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       name  binom_p_val  hyper_p_val  \\\n",
       "113  positive regulation of gtpase activity  2.67021e-03  5.01449e-02   \n",
       "\n",
       "         p_binom      k_binom          id  \n",
       "113  2.86761e-02  8.00000e+00  GO:0043547  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loc_gpy[loc_gpy[\"id\"] == \"GO:0043547\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the online webserver we find the following results:\n",
    "\n",
    "* binom_p_val : 6.6206e-3\n",
    "* fold enr : 2.3810\n",
    "* observed hit : 11\n",
    "* coverage : 10.58%\n",
    "* hyper_p_val : 3.6335e-2 \n",
    "* fold enr : 3.1974\n",
    "* obs gene hit : 4\n",
    "* total gene set : 405\n",
    "* hyper coverage : 6.90%\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('gpy')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "610c73e5f1997caa668091b43a573fbafba1e2b96e55fbc05d56c58d21b90949"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
